{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define a `dict` so we can convert 3-letter amino acids to 1-letter codes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "amino_acid_lookup = {'Cys': 'C', 'Asp': 'D', 'Ser': 'S', 'Gln': 'Q', 'Lys': 'K', 'Ile': 'I', 'Pro': 'P', 'Thr': 'T', 'Phe': 'F', 'Asn': 'N', 'Gly': 'G', 'His': 'H', 'Leu': 'L', 'Arg': 'R', 'Trp': 'W', 'Ala': 'A', 'Val':'V', 'Glu': 'E', 'Tyr': 'Y', 'Met': 'M', 'stop': '!'}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load in the amino acid sequence of version 3 of H37Rv / NC_000962.\n",
    "\n",
    "Retrieved from mycobrowser.epfl.ch on 6 Jan 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/reference/NC_000962.3_pncA_aminoacids.fasta','r') as INPUT:\n",
    "    INPUT.readline()\n",
    "    pncA_amino_acid_sequence = INPUT.readline()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test/train dataset\n",
    "\n",
    "This is derived from three sources\n",
    "1. an in vitro mutagenesis study by Yadon et al. (Yadon2017)\n",
    "2. the NEJM resistance catalogue (NEJM2018)\n",
    "3. the WHO resistance catalogue (WHO2021)\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Yadon2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fowler/Library/Python/3.10/lib/python/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNDER_REPRESENTED</th>\n",
       "      <th>IN_VITRO_PHENOTYPE</th>\n",
       "      <th>IN_VIVO_PHENOTYPE</th>\n",
       "      <th>YADON_RESULT</th>\n",
       "      <th>MIOTTO2014</th>\n",
       "      <th>WALKER2015</th>\n",
       "      <th>IS_NONSYNONYMOUS</th>\n",
       "      <th>IS_SYNONYMOUS</th>\n",
       "      <th>CONTAINS_STOP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M1I</th>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1L</th>\n",
       "      <td>False</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1V</th>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          UNDER_REPRESENTED IN_VITRO_PHENOTYPE IN_VIVO_PHENOTYPE YADON_RESULT  \\\n",
       "MUTATION                                                                        \n",
       "M1I                   False               None                 R            R   \n",
       "M1L                   False                  R                 R            R   \n",
       "M1V                   False               None                 S            S   \n",
       "\n",
       "          MIOTTO2014  WALKER2015  IS_NONSYNONYMOUS  IS_SYNONYMOUS  \\\n",
       "MUTATION                                                            \n",
       "M1I            False       False              True          False   \n",
       "M1L            False       False              True          False   \n",
       "M1V            False       False              True          False   \n",
       "\n",
       "          CONTAINS_STOP  \n",
       "MUTATION                 \n",
       "M1I               False  \n",
       "M1L               False  \n",
       "M1V               False  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "YADON = pandas.read_excel('./data/source-data/41467_2017_721_MOESM4_ESM.xlsx')\n",
    "\n",
    "# rename the columns to not contain spaces\n",
    "YADON.rename(columns={  'Substitution':'MUTATION',\n",
    "                        'Under-Represented?': 'UNDER_REPRESENTED',\n",
    "                        'Catalogue': 'CATALOGUE',\n",
    "                        'Miotto et. al.': 'MIOTTO2014',\n",
    "                        'Walker et. al.': 'WALKER2015'}, inplace=True)\n",
    "\n",
    "# make a Boolean IS_NONSYN columns\n",
    "def define_nonsyn(row):\n",
    "    IS_NONSYN = False\n",
    "    IS_SYN = False\n",
    "    CONTAINS_STOP = False\n",
    "    if row['MUTATION'][-1]=='*':\n",
    "        CONTAINS_STOP=True\n",
    "    if row['MUTATION'][0]==row['MUTATION'][-1]:\n",
    "        IS_SYN=True\n",
    "    elif row['MUTATION'][0]!=row['MUTATION'][-1]:\n",
    "        IS_NONSYN=True\n",
    "    return pandas.Series([IS_NONSYN,IS_SYN,CONTAINS_STOP])\n",
    "\n",
    "YADON[['IS_NONSYNONYMOUS','IS_SYNONYMOUS','CONTAINS_STOP']] = YADON.apply(define_nonsyn, axis=1)       \n",
    "YADON['UNDER_REPRESENTED'] = YADON.UNDER_REPRESENTED=='Y'\n",
    "\n",
    "# replace the NaNs with zeros and then convert to True/False                        \n",
    "for i in ['in vitro Resistant', 'in vitro Susceptible', 'in vivo Resistant', 'in vivo Susceptible','MIOTTO2014','WALKER2015']:\n",
    "    YADON[i] = YADON[i].fillna(0).astype('bool')                        \n",
    "\n",
    "# combine the one-hot encoded columns\n",
    "def assign_phenotypes(row, string):\n",
    "    if row[string+' Resistant'] and not row[string+' Susceptible']:\n",
    "        return('R')\n",
    "    elif not row[string+' Resistant'] and row[string+' Susceptible']:\n",
    "        return('S')\n",
    "    else:\n",
    "        return(None)\n",
    "\n",
    "YADON['IN_VITRO_PHENOTYPE'] = YADON.apply(assign_phenotypes, args=('in vitro',), axis=1)\n",
    "YADON['IN_VIVO_PHENOTYPE'] = YADON.apply(assign_phenotypes, args=('in vivo',), axis=1)\n",
    "\n",
    "# drop the one-hot columns\n",
    "YADON.drop(columns=['in vitro Resistant', 'in vitro Susceptible', 'in vivo Resistant', 'in vivo Susceptible'], inplace=True)\n",
    "\n",
    "def interpet_catalogue(row):\n",
    "    if row.CATALOGUE=='E':\n",
    "        return('R')\n",
    "    elif row.CATALOGUE=='D':\n",
    "        return('S')\n",
    "\n",
    "YADON['YADON_RESULT'] = YADON.apply(interpet_catalogue, axis=1)       \n",
    "\n",
    "def replace_stop_codon(row):\n",
    "    if \"*\" in row.MUTATION:\n",
    "        return row.MUTATION.replace(\"*\",\"!\")\n",
    "    else:\n",
    "        return row.MUTATION\n",
    "\n",
    "YADON['MUTATION'] = YADON.apply(replace_stop_codon, axis=1)\n",
    "\n",
    "# reorder the dataset\n",
    "YADON = YADON[['MUTATION', 'UNDER_REPRESENTED', 'IN_VITRO_PHENOTYPE', 'IN_VIVO_PHENOTYPE', 'YADON_RESULT', 'MIOTTO2014',\n",
    "       'WALKER2015', 'IS_NONSYNONYMOUS','IS_SYNONYMOUS','CONTAINS_STOP']]\n",
    "\n",
    "YADON = YADON[(~YADON.UNDER_REPRESENTED) & (YADON.YADON_RESULT.notna())]\n",
    "YADON.set_index('MUTATION', inplace=True)\n",
    "\n",
    "YADON[['YADON_RESULT']].to_csv('data/catalogues/ds-yadon2017.csv', index=True)\n",
    "\n",
    "YADON[:3]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## NEJM2018"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>GENBANK_REFERENCE</th>\n",
       "      <th>CATALOGUE_NAME</th>\n",
       "      <th>CATALOGUE_VERSION</th>\n",
       "      <th>CATALOGUE_GRAMMAR</th>\n",
       "      <th>PREDICTION_VALUES</th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>SOURCE</th>\n",
       "      <th>EVIDENCE</th>\n",
       "      <th>OTHER</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td>469</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>NEJM2018</td>\n",
       "      <td>v1.1</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RUS</td>\n",
       "      <td>PZA</td>\n",
       "      <td>-3_indel</td>\n",
       "      <td>S</td>\n",
       "      <td>{\"DOI\": \"10.1056/NEJMoa1800474\"}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{\"LITERATURE_SOURCE\": \"Walker2015\", \"LITERATUR...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>pncA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>472</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>NEJM2018</td>\n",
       "      <td>v1.1</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RUS</td>\n",
       "      <td>PZA</td>\n",
       "      <td>145_indel</td>\n",
       "      <td>S</td>\n",
       "      <td>{\"DOI\": \"10.1056/NEJMoa1800474\"}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{\"LITERATURE_SOURCE\": \"Walker2015\", \"LITERATUR...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>pncA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>473</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>NEJM2018</td>\n",
       "      <td>v1.1</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RUS</td>\n",
       "      <td>PZA</td>\n",
       "      <td>185_indel</td>\n",
       "      <td>R</td>\n",
       "      <td>{\"DOI\": \"10.1056/NEJMoa1800474\"}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{\"LITERATURE_SOURCE\": \"Walker2015\", \"LITERATUR...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>pncA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0 GENBANK_REFERENCE CATALOGUE_NAME CATALOGUE_VERSION  \\\n",
       "450         469       NC_000962.3       NEJM2018              v1.1   \n",
       "451         472       NC_000962.3       NEJM2018              v1.1   \n",
       "452         473       NC_000962.3       NEJM2018              v1.1   \n",
       "\n",
       "    CATALOGUE_GRAMMAR PREDICTION_VALUES DRUG   MUTATION PREDICTION  \\\n",
       "450             GARC1               RUS  PZA   -3_indel          S   \n",
       "451             GARC1               RUS  PZA  145_indel          S   \n",
       "452             GARC1               RUS  PZA  185_indel          R   \n",
       "\n",
       "                               SOURCE EVIDENCE  \\\n",
       "450  {\"DOI\": \"10.1056/NEJMoa1800474\"}       {}   \n",
       "451  {\"DOI\": \"10.1056/NEJMoa1800474\"}       {}   \n",
       "452  {\"DOI\": \"10.1056/NEJMoa1800474\"}       {}   \n",
       "\n",
       "                                                 OTHER  IS_SNP  IN_CDS  GENE  \n",
       "450  {\"LITERATURE_SOURCE\": \"Walker2015\", \"LITERATUR...   False    True  pncA  \n",
       "451  {\"LITERATURE_SOURCE\": \"Walker2015\", \"LITERATUR...   False    True  pncA  \n",
       "452  {\"LITERATURE_SOURCE\": \"Walker2015\", \"LITERATUR...   False    True  pncA  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nejm2018 = pandas.read_csv('data/source-data/NC_000962.3_NEJM2018_v1.1_GARC1_RUS.csv')\n",
    "\n",
    "nejm2018 = nejm2018[(nejm2018.DRUG=='PZA') & (nejm2018.PREDICTION.isin(['R','S'])) & (~nejm2018.MUTATION.str.contains('*', regex=False))]\n",
    "\n",
    "def infer_type(row):\n",
    "    mut = row.MUTATION\n",
    "    cols = mut.split('@')\n",
    "    IS_SNP = True\n",
    "    IN_CDS = True\n",
    "    if 'ins' in mut:\n",
    "        IS_SNP=False\n",
    "    elif 'del' in mut:\n",
    "        IS_SNP=False\n",
    "    elif '-' in mut:\n",
    "        IN_CDS=False\n",
    "    return pandas.Series([IS_SNP, IN_CDS, cols[0], cols[1]])        \n",
    "\n",
    "nejm2018[['IS_SNP', 'IN_CDS', 'GENE', 'MUTATION']] = nejm2018.apply(infer_type, axis=1)    \n",
    "\n",
    "nejm2018[['MUTATION', 'PREDICTION']].to_csv('data/catalogues/ds-nejm2018.csv', index=False)\n",
    "\n",
    "nejm2018[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## WHO2021"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENBANK_REFERENCE</th>\n",
       "      <th>CATALOGUE_NAME</th>\n",
       "      <th>CATALOGUE_VERSION</th>\n",
       "      <th>CATALOGUE_GRAMMAR</th>\n",
       "      <th>PREDICTION_VALUES</th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>SOURCE</th>\n",
       "      <th>EVIDENCE</th>\n",
       "      <th>OTHER</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>WHO-UCN-GTB-PCI-2021.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RUS</td>\n",
       "      <td>PZA</td>\n",
       "      <td>-5_del_g</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{\"Present_SOLO_R\": 9, \"Present_SOLO_SR\": 10, \"...</td>\n",
       "      <td>{\"FINAL_CONFIDENCE_GRADING\": \"1) Assoc w R\"}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>pncA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>WHO-UCN-GTB-PCI-2021.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RUS</td>\n",
       "      <td>PZA</td>\n",
       "      <td>108_ins_tacctggc</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{\"Present_SOLO_R\": 1, \"Present_SOLO_SR\": 1, \"P...</td>\n",
       "      <td>{\"FINAL_CONFIDENCE_GRADING\": \"2) Assoc w R - I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>pncA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>WHO-UCN-GTB-PCI-2021.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RUS</td>\n",
       "      <td>PZA</td>\n",
       "      <td>116_del_cggactaccatcacgtc</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{\"Present_SOLO_R\": 1, \"Present_SOLO_SR\": 1, \"P...</td>\n",
       "      <td>{\"FINAL_CONFIDENCE_GRADING\": \"2) Assoc w R - I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>pncA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    GENBANK_REFERENCE          CATALOGUE_NAME  CATALOGUE_VERSION  \\\n",
       "228       NC_000962.3  WHO-UCN-GTB-PCI-2021.7                1.0   \n",
       "229       NC_000962.3  WHO-UCN-GTB-PCI-2021.7                1.0   \n",
       "230       NC_000962.3  WHO-UCN-GTB-PCI-2021.7                1.0   \n",
       "\n",
       "    CATALOGUE_GRAMMAR PREDICTION_VALUES DRUG                   MUTATION  \\\n",
       "228             GARC1               RUS  PZA                   -5_del_g   \n",
       "229             GARC1               RUS  PZA           108_ins_tacctggc   \n",
       "230             GARC1               RUS  PZA  116_del_cggactaccatcacgtc   \n",
       "\n",
       "    PREDICTION SOURCE                                           EVIDENCE  \\\n",
       "228          R     {}  {\"Present_SOLO_R\": 9, \"Present_SOLO_SR\": 10, \"...   \n",
       "229          R     {}  {\"Present_SOLO_R\": 1, \"Present_SOLO_SR\": 1, \"P...   \n",
       "230          R     {}  {\"Present_SOLO_R\": 1, \"Present_SOLO_SR\": 1, \"P...   \n",
       "\n",
       "                                                 OTHER  Unnamed: 0  IS_SNP  \\\n",
       "228       {\"FINAL_CONFIDENCE_GRADING\": \"1) Assoc w R\"}         NaN   False   \n",
       "229  {\"FINAL_CONFIDENCE_GRADING\": \"2) Assoc w R - I...         NaN   False   \n",
       "230  {\"FINAL_CONFIDENCE_GRADING\": \"2) Assoc w R - I...         NaN   False   \n",
       "\n",
       "     IN_CDS  GENE  \n",
       "228    True  pncA  \n",
       "229    True  pncA  \n",
       "230    True  pncA  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "who = pandas.read_csv('data/source-data/NC_000962.3_WHO-UCN-GTB-PCI-2021.7_v1.0_GARC1_RUS.csv')\n",
    "\n",
    "who = who[(who.DRUG=='PZA') & (who.PREDICTION.isin(['R','S'])) & (~who.MUTATION.str.contains('*', regex=False))]\n",
    "\n",
    "def infer_type(row):\n",
    "    mut = row.MUTATION\n",
    "    cols = mut.split('@')\n",
    "    IS_SNP = True\n",
    "    IN_CDS = True\n",
    "    if 'ins' in mut:\n",
    "        IS_SNP=False\n",
    "    elif 'del' in mut:\n",
    "        IS_SNP=False\n",
    "    elif '-' in mut:\n",
    "        IN_CDS=False\n",
    "    return pandas.Series([IS_SNP, IN_CDS, cols[0], cols[1]])        \n",
    "\n",
    "who[['IS_SNP', 'IN_CDS', 'GENE', 'MUTATION']] = who.apply(infer_type, axis=1)    \n",
    "\n",
    "who[['MUTATION', 'PREDICTION']].to_csv('data/catalogues/ds-who2021.csv', index=False)\n",
    "\n",
    "who[:3]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validation dataset\n",
    "\n",
    "This is also derived from three sources"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CRyPTIC2021"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    15991\n",
       "True      6851\n",
       "Name: HAS_MUTATION, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DST_CORE = pandas.read_csv('data/source-data/cryptic1-dst.csv')\n",
    "DST_CORE.set_index('UNIQUEID', inplace=True)\n",
    "\n",
    "MUTATIONS_CORE = pandas.read_csv('data/source-data/cryptic1-mutations.csv')\n",
    "MUTATIONS_CORE.set_index('UNIQUEID', inplace=True)\n",
    "samples_with_mutations=set(MUTATIONS_CORE.index)\n",
    "\n",
    "DST_CORE.loc[DST_CORE.index.isin(samples_with_mutations), 'HAS_MUTATION']=True\n",
    "DST_CORE.HAS_MUTATION.fillna(False,inplace=True)\n",
    "DST_CORE.HAS_MUTATION.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 3,549 samples with one or more mutations in pncA (and 19,293 that do not). Mutation here means a non-synonymous mutation (missense and nonsense), an assumed promoter mutation or a detected insertion or deletion."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    6622\n",
       "2     228\n",
       "3       1\n",
       "Name: N_MUTATIONS, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS_CORE.reset_index(inplace=True)\n",
    "MUTATIONS_CORE_COUNTS=MUTATIONS_CORE[['UNIQUEID','SITEID']].groupby('UNIQUEID').count()\n",
    "MUTATIONS_CORE.set_index('UNIQUEID',inplace=True)\n",
    "MUTATIONS_CORE_COUNTS.rename(columns={'SITEID':'N_MUTATIONS'},inplace=True)\n",
    "MUTATIONS_CORE_COUNTS\n",
    "MUTATIONS_CORE_COUNTS.N_MUTATIONS.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>...</th>\n",
       "      <th>IS_FILTER_PASS</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>SOURCE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.00.subj.1000347.lab.H111540004.iso.1</th>\n",
       "      <td>pncA</td>\n",
       "      <td>Q141P</td>\n",
       "      <td>141.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cag</td>\n",
       "      <td>ccg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>00</td>\n",
       "      <td>1</td>\n",
       "      <td>R</td>\n",
       "      <td>NEJM2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.00.subj.1000595.lab.H123460044.iso.1</th>\n",
       "      <td>pncA</td>\n",
       "      <td>D12A</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gac</td>\n",
       "      <td>gcc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>00</td>\n",
       "      <td>1</td>\n",
       "      <td>S</td>\n",
       "      <td>NEJM2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.00.subj.1004213.lab.H111060034.iso.1</th>\n",
       "      <td>pncA</td>\n",
       "      <td>392_indel</td>\n",
       "      <td>392.0</td>\n",
       "      <td>131.0</td>\n",
       "      <td>2288850.0</td>\n",
       "      <td>392.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>2.0</td>\n",
       "      <td>392_ins</td>\n",
       "      <td>392_ins_2</td>\n",
       "      <td>00</td>\n",
       "      <td>0</td>\n",
       "      <td>R</td>\n",
       "      <td>NEJM2018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           GENE   MUTATION  POSITION  \\\n",
       "UNIQUEID                                                               \n",
       "site.00.subj.1000347.lab.H111540004.iso.1  pncA      Q141P     141.0   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1  pncA       D12A      12.0   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1  pncA  392_indel     392.0   \n",
       "\n",
       "                                           AMINO_ACID_NUMBER  GENOME_INDEX  \\\n",
       "UNIQUEID                                                                     \n",
       "site.00.subj.1000347.lab.H111540004.iso.1              141.0           NaN   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1               12.0           NaN   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1              131.0     2288850.0   \n",
       "\n",
       "                                           NUCLEOTIDE_NUMBER  REF  ALT  \\\n",
       "UNIQUEID                                                                 \n",
       "site.00.subj.1000347.lab.H111540004.iso.1                NaN  cag  ccg   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1                NaN  gac  gcc   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1              392.0  NaN  NaN   \n",
       "\n",
       "                                           IS_SNP  IS_INDEL  ...  \\\n",
       "UNIQUEID                                                     ...   \n",
       "site.00.subj.1000347.lab.H111540004.iso.1    True     False  ...   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1    True     False  ...   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1   False      True  ...   \n",
       "\n",
       "                                           IS_FILTER_PASS  ELEMENT_TYPE  \\\n",
       "UNIQUEID                                                                  \n",
       "site.00.subj.1000347.lab.H111540004.iso.1            True          GENE   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1            True          GENE   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1            True          GENE   \n",
       "\n",
       "                                           MUTATION_TYPE  INDEL_LENGTH  \\\n",
       "UNIQUEID                                                                 \n",
       "site.00.subj.1000347.lab.H111540004.iso.1            AAM           NaN   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1            AAM           NaN   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1          INDEL           2.0   \n",
       "\n",
       "                                           INDEL_1    INDEL_2  SITEID  \\\n",
       "UNIQUEID                                                                \n",
       "site.00.subj.1000347.lab.H111540004.iso.1      NaN        NaN      00   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1      NaN        NaN      00   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1  392_ins  392_ins_2      00   \n",
       "\n",
       "                                          NUMBER_NUCLEOTIDE_CHANGES PHENOTYPE  \\\n",
       "UNIQUEID                                                                        \n",
       "site.00.subj.1000347.lab.H111540004.iso.1                         1         R   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1                         1         S   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1                         0         R   \n",
       "\n",
       "                                             SOURCE  \n",
       "UNIQUEID                                             \n",
       "site.00.subj.1000347.lab.H111540004.iso.1  NEJM2018  \n",
       "site.00.subj.1000595.lab.H123460044.iso.1  NEJM2018  \n",
       "site.00.subj.1004213.lab.H111060034.iso.1  NEJM2018  \n",
       "\n",
       "[3 rows x 26 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DST_CORE=DST_CORE.join(MUTATIONS_CORE_COUNTS,how='left')\n",
    "DST_CORE.N_MUTATIONS.fillna(0,inplace=True) \n",
    "DST_CORE['IS_SOLO']=DST_CORE.N_MUTATIONS==1\n",
    "SOLOS=MUTATIONS_CORE.join(DST_CORE[DST_CORE.IS_SOLO][['PHENOTYPE', 'SOURCE']],how='inner')\n",
    "SOLOS[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-2_indel</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-32_indel</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-3_indel</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-4_indel</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-9_indel</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g-9a</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t-10c</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t-12c</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t-7c</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t-7g</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>537 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "PHENOTYPE                 R   S\n",
       "MUTATION  IS_SNP IN_CDS        \n",
       "-2_indel  False  False    0  12\n",
       "-32_indel False  False    0   1\n",
       "-3_indel  False  False    2   0\n",
       "-4_indel  False  False   10   1\n",
       "-9_indel  False  False    0   1\n",
       "...                      ..  ..\n",
       "g-9a      True   False    0   1\n",
       "t-10c     True   False    0   2\n",
       "t-12c     True   False    8   1\n",
       "t-7c      True   False    5   0\n",
       "t-7g      True   False    1   0\n",
       "\n",
       "[537 rows x 2 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SOLO_MUTATIONS_CROSSTAB=pandas.crosstab([SOLOS.MUTATION,SOLOS.IS_SNP,SOLOS.IN_CDS],SOLOS.PHENOTYPE,margins=False,margins_name='TOTAL')\n",
    "SOLO_MUTATIONS_CROSSTAB.to_csv('data/clinical-samples/ds-cryptic2021.csv')\n",
    "SOLO_MUTATIONS_CROSSTAB"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Miotto2014 data\n",
    "\n",
    "> Miotto, P., Cabibbe, A. M., Feuerriegel, S., Casali, N., Drobniewski, F., Rodionova, Y., Bakonyte, D., Stakenas, P., Pimkina, E., Augustynowicz-Kopeć, E., Degano, M., Ambrosi, A., Hoffner, S., Mansjö, M., Werngren, J., Rüsch-Gerdes, S., Niemann, S., & Cirillo, D. M. (2014). Mycobacterium tuberculosis pyrazinamide resistance determinants: a multicenter study. mBio, 5(5), e01819-14. https://doi.org/10.1128/mBio.01819-14\n",
    "\n",
    "Load the raw Excel file from the supplement of this paper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Isolate n</th>\n",
       "      <th>From</th>\n",
       "      <th>Other (SIRE)</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>pncA nt</th>\n",
       "      <th>pncA aa</th>\n",
       "      <th>Δ free energy (pH 6)</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Structure analysis</th>\n",
       "      <th>Enzymatic activity</th>\n",
       "      <th>...</th>\n",
       "      <th>Structure</th>\n",
       "      <th>Free energy</th>\n",
       "      <th>Summary structure+energy</th>\n",
       "      <th>Lineage.1</th>\n",
       "      <th>p.S</th>\n",
       "      <th>p.R</th>\n",
       "      <th>Category</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Location_12</td>\n",
       "      <td>none</td>\n",
       "      <td>S</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>West African 1</td>\n",
       "      <td>0.841935</td>\n",
       "      <td>0.158065</td>\n",
       "      <td>B</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Location_12</td>\n",
       "      <td>multi</td>\n",
       "      <td>S</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>LAM</td>\n",
       "      <td>0.841935</td>\n",
       "      <td>0.158065</td>\n",
       "      <td>B</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Location_12</td>\n",
       "      <td>mono</td>\n",
       "      <td>S</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>Sierra Leone-1</td>\n",
       "      <td>0.841935</td>\n",
       "      <td>0.158065</td>\n",
       "      <td>B</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Isolate n         From Other (SIRE) PHENOTYPE pncA nt pncA aa  \\\n",
       "0          1  Location_12         none         S      WT      WT   \n",
       "1          2  Location_12        multi         S      WT      WT   \n",
       "2          3  Location_12         mono         S      WT      WT   \n",
       "\n",
       "  Δ free energy (pH 6) Notes Structure analysis Enzymatic activity  ...  \\\n",
       "0                    -   NaN                NaN                 na  ...   \n",
       "1                    -   NaN                NaN                 na  ...   \n",
       "2                    -   NaN                NaN                 na  ...   \n",
       "\n",
       "  Structure Free energy Summary structure+energy       Lineage.1       p.S  \\\n",
       "0         S           S                        S  West African 1  0.841935   \n",
       "1         S           S                        S             LAM  0.841935   \n",
       "2         S           S                        S  Sierra Leone-1  0.841935   \n",
       "\n",
       "        p.R Category  MUTATION  IS_SNP IN_CDS  \n",
       "0  0.158065        B      None    None   None  \n",
       "1  0.158065        B      None    None   None  \n",
       "2  0.158065        B      None    None   None  \n",
       "\n",
       "[3 rows x 23 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MIOTTO2014 = pandas.read_excel('./data/source-data/mbo005142032st1.xlsx', header=2)\n",
    "\n",
    "def lookup_mutation(row):\n",
    "\n",
    "    mutation = None\n",
    "    IS_SNP = None\n",
    "    IN_CDS = None\n",
    "    \n",
    "    # skipping rows with a comma skips samples with multiple mutations\n",
    "    if row['pncA aa'] != 'WT' and ',' not in row['pncA nt'] and ',' not in row['pncA aa'] and row['pncA nt'] != 'del >200 nt':\n",
    "        if row['pncA aa'][:3] not in amino_acid_lookup.keys():\n",
    "            if 'promoter' in row['pncA aa']: # or 'shift' not in row['pncA aa']:\n",
    "                IN_CDS = False\n",
    "                if row['pncA nt'][:3]  in ['Ins', 'Del']: \n",
    "                    IS_SNP = False\n",
    "                    mut = row['pncA nt']\n",
    "                    pos = mut.split(' ')[0][3:]\n",
    "                    mutation = ('indel_'+str(int(pos)))\n",
    "                else:\n",
    "                    IS_SNP = True\n",
    "                    mut = row['pncA nt']\n",
    "                    assert mut[0] in ['a','t','c','g'], mut\n",
    "                    assert mut[-1] in ['a','t','c','g'], mut\n",
    "                    mutation = (row['pncA nt'])\n",
    "        else:\n",
    "            mut = row['pncA aa']\n",
    "            IS_SNP = True\n",
    "            IN_CDS = True\n",
    "            if 'stop' in mut:\n",
    "                if mut[:4] == 'stop':\n",
    "                    mutation = (amino_acid_lookup[mut[:4]] + mut[4:-3] + amino_acid_lookup[mut[-3:]])\n",
    "                else:\n",
    "                    mutation = (amino_acid_lookup[mut[:3]] + mut[3:-4] + amino_acid_lookup[mut[-4:]])\n",
    "            else:\n",
    "                ref = amino_acid_lookup[mut[:3]]\n",
    "                pos = mut[3:-3]\n",
    "                alt = amino_acid_lookup[mut[-3:]]\n",
    "                assert pncA_amino_acid_sequence[int(pos)-1] == ref, mut\n",
    "\n",
    "                mutation = (ref + pos + alt)\n",
    "    return pandas.Series([mutation, IS_SNP, IN_CDS])\n",
    "    \n",
    "MIOTTO2014[['MUTATION', 'IS_SNP', 'IN_CDS']] = MIOTTO2014.apply(lookup_mutation, axis=1)\n",
    "\n",
    "MIOTTO2014.rename(columns = {'PZA': 'PHENOTYPE'}, inplace=True)\n",
    "\n",
    "MIOTTO2014[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A102P</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A102R</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A102T</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A102V</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A134V</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g-13t</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>indel_-3</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>indel_-5</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t-7c</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t-7g</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>199 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "PHENOTYPE               R  S\n",
       "MUTATION IS_SNP IN_CDS      \n",
       "A102P    True   True    0  1\n",
       "A102R    True   True    0  1\n",
       "A102T    True   True    1  0\n",
       "A102V    True   True    1  1\n",
       "A134V    True   True    2  0\n",
       "...                    .. ..\n",
       "g-13t    True   False   0  1\n",
       "indel_-3 False  False   1  1\n",
       "indel_-5 False  False   2  0\n",
       "t-7c     True   False   4  0\n",
       "t-7g     True   False   1  0\n",
       "\n",
       "[199 rows x 2 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MIOTTO2014_MUTATIONS_SUMMARY=pandas.crosstab([MIOTTO2014.MUTATION,MIOTTO2014.IS_SNP,MIOTTO2014.IN_CDS],MIOTTO2014.PHENOTYPE,margins=False)\n",
    "MIOTTO2014_MUTATIONS_SUMMARY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are a total of 755 samples in the Miotto2014 dataset\n"
     ]
    }
   ],
   "source": [
    "n_miotto = MIOTTO2014_MUTATIONS_SUMMARY.R.sum() + MIOTTO2014_MUTATIONS_SUMMARY.S.sum()\n",
    "print(\"There are a total of \" + str(n_miotto) + \" samples in the Miotto2014 dataset\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 191 non-synoymous mutations in 704 samples\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A102P</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A102R</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A102T</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PHENOTYPE               R  S\n",
       "MUTATION IS_SNP IN_CDS      \n",
       "A102P    True   True    0  1\n",
       "A102R    True   True    0  1\n",
       "A102T    True   True    1  0"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MIOTTO2014_MUTATIONS_SUMMARY.reset_index(inplace=True)\n",
    "n_miotto_nonsyn = MIOTTO2014_MUTATIONS_SUMMARY[(MIOTTO2014_MUTATIONS_SUMMARY.IS_SNP) & (MIOTTO2014_MUTATIONS_SUMMARY.IN_CDS)]\n",
    "print(\"There are \" + str(len(n_miotto_nonsyn)) + \" non-synoymous mutations in \" + str(n_miotto_nonsyn.R.sum() + n_miotto_nonsyn.S.sum()) + \" samples\")\n",
    "MIOTTO2014_MUTATIONS_SUMMARY.set_index(['MUTATION', 'IS_SNP', 'IN_CDS'], inplace=True)\n",
    "MIOTTO2014_MUTATIONS_SUMMARY[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "MIOTTO2014_MUTATIONS_SUMMARY.to_csv('data/clinical-samples/ds-miotto2014.csv',index=True)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Whitfield2015\n",
    "\n",
    "> Whitfield, M. G., Soeters, H. M., Warren, R. M., York, T., Sampson, S. L., Streicher, E. M., Van Helden, P. D., & Van Rie, A. (2015). A global perspective on pyrazinamide resistance: Systematic review and meta-analysis. PLoS ONE, 10(7), 1–16. https://doi.org/10.1371/journal.pone.0133869\n",
    "\n",
    "Load in the Whitfield data -- this was copied from `zjm999094564so1.pdf`, missing out any promoter/indels, into an Excel sheet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>I5T</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I6L</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V9V</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N11D</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D12A</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>26</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T177P</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L182S</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L182W</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!187G</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!187R</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         R    S\n",
       "MUTATION IS_SNP IN_CDS         \n",
       "I5T      True   True     0    1\n",
       "I6L      True   True     0  128\n",
       "V9V      True   True     0    1\n",
       "N11D     True   True     0    1\n",
       "D12A     True   True    26    2\n",
       "...                     ..  ...\n",
       "T177P    True   True     1    1\n",
       "L182S    True   True     2    1\n",
       "L182W    True   True     0    1\n",
       "!187G    True   True     1    1\n",
       "!187R    True   True     0    1\n",
       "\n",
       "[65 rows x 2 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "WHITFIELD2015 = pandas.read_excel('./data/source-data/zjm999094564so1.xlsx', header=1)\n",
    "\n",
    "def parse_whitfield2015(row):\n",
    "\n",
    "    mut = row.MUTATION.rstrip()\n",
    "    IS_SNP = True\n",
    "    IN_CDS = True\n",
    "\n",
    "    if mut[:4]== 'stop':\n",
    "        ref = amino_acid_lookup[mut[:4]]\n",
    "        pos = mut[4:-3]\n",
    "        alt = amino_acid_lookup[mut[-3:]]\n",
    "    elif mut[-4:] == 'stop':\n",
    "        ref = amino_acid_lookup[mut[:3]]\n",
    "        pos = mut[3:-4]\n",
    "        alt = amino_acid_lookup[mut[-4:]]\n",
    "    else:\n",
    "        ref = amino_acid_lookup[mut[:3]]\n",
    "        pos = mut[3:-3]\n",
    "        alt = amino_acid_lookup[mut[-3:]]\n",
    "\n",
    "    mutation = ref + pos + alt\n",
    "    if int(pos) <= 186:\n",
    "        assert pncA_amino_acid_sequence[int(pos)-1] == ref, mut\n",
    "    else:\n",
    "        assert '!' == ref , mut\n",
    "\n",
    "    \n",
    "    return(pandas.Series([mutation, IS_SNP, IN_CDS]))\n",
    "\n",
    "WHITFIELD2015[['MUTATION', 'IS_SNP', 'IN_CDS']] = WHITFIELD2015.apply(parse_whitfield2015, axis=1)\n",
    "# WHITFIELD2015.rename(columns={'R': 'Rw', 'S': 'Sw'}, inplace=True)\n",
    "WHITFIELD2015 = WHITFIELD2015[WHITFIELD2015.MUTATION.notna()][['MUTATION', 'IS_SNP', 'IN_CDS', 'R', 'S']]\n",
    "WHITFIELD2015.set_index(['MUTATION', 'IS_SNP', 'IN_CDS'], inplace=True)\n",
    "WHITFIELD2015.to_csv('data/clinical-samples/ds-whitfield2015.csv')\n",
    "WHITFIELD2015"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'Rw'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [135], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39mThere are \u001b[39m\u001b[39m\"\u001b[39m \u001b[39m+\u001b[39m \u001b[39mstr\u001b[39m(\u001b[39mlen\u001b[39m(WHITFIELD2015)) \u001b[39m+\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m non-synonymous mutations in \u001b[39m\u001b[39m\"\u001b[39m \u001b[39m+\u001b[39m \u001b[39mstr\u001b[39m(WHITFIELD2015\u001b[39m.\u001b[39;49mRw\u001b[39m.\u001b[39msum() \u001b[39m+\u001b[39m WHITFIELD2015\u001b[39m.\u001b[39mSw\u001b[39m.\u001b[39msum()) \u001b[39m+\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m samples in the Whitfield2015 dataset.\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m~/Library/Python/3.10/lib/python/site-packages/pandas/core/generic.py:5902\u001b[0m, in \u001b[0;36mNDFrame.__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5895\u001b[0m \u001b[39mif\u001b[39;00m (\n\u001b[1;32m   5896\u001b[0m     name \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_internal_names_set\n\u001b[1;32m   5897\u001b[0m     \u001b[39mand\u001b[39;00m name \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_metadata\n\u001b[1;32m   5898\u001b[0m     \u001b[39mand\u001b[39;00m name \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_accessors\n\u001b[1;32m   5899\u001b[0m     \u001b[39mand\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_info_axis\u001b[39m.\u001b[39m_can_hold_identifiers_and_holds_name(name)\n\u001b[1;32m   5900\u001b[0m ):\n\u001b[1;32m   5901\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m[name]\n\u001b[0;32m-> 5902\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mobject\u001b[39;49m\u001b[39m.\u001b[39;49m\u001b[39m__getattribute__\u001b[39;49m(\u001b[39mself\u001b[39;49m, name)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'Rw'"
     ]
    }
   ],
   "source": [
    "print(\"There are \" + str(len(WHITFIELD2015)) + \" non-synonymous mutations in \" + str(WHITFIELD2015.Rw.sum() + WHITFIELD2015.Sw.sum()) + \" samples in the Whitfield2015 dataset.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "7a792fcb311f9eb9f3c1b942a8c87ada8484712b89b670347c16a1088e0a1f69"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
