{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# `01-parse-original-data`\n",
    "\n",
    "This notebook reads the original (if possible) source of the data, such as an Excel sheet provided as Supplementary Information to the cited reference. Since all the data sources use slightly different conventions there is a substantial amount of cleaning to get each into the same, simple format that the next notebooks then consume."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "\n",
    "pandas.options.display.max_columns=999"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define a `dict` so we can convert 3-letter amino acids to 1-letter codes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "amino_acid_lookup = {'Cys': 'C', 'Asp': 'D', 'Ser': 'S', 'Gln': 'Q', 'Lys': 'K', 'Ile': 'I', 'Pro': 'P', 'Thr': 'T', 'Phe': 'F', 'Asn': 'N', 'Gly': 'G', 'His': 'H', 'Leu': 'L', 'Arg': 'R', 'Trp': 'W', 'Ala': 'A', 'Val':'V', 'Glu': 'E', 'Tyr': 'Y', 'Met': 'M', 'stop': '!'}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load in the amino acid sequence of version 3 of H37Rv / NC_000962. Retrieved from mycobrowser.epfl.ch on 6 Jan 2023."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/reference/NC_000962.3_pncA_aminoacids.fasta','r') as INPUT:\n",
    "    INPUT.readline()\n",
    "    pncA_amino_acid_sequence = INPUT.readline()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Test/Train dataset\n",
    "\n",
    "This is derived from three sources\n",
    "(a). an in vitro mutagenesis study by Yadon et al. (Yadon2017)\n",
    "(b). the NEJM resistance catalogue (NEJM2018)\n",
    "(c). the WHO resistance catalogue (WHO2021)\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1 (a). Yadon2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fowler/Library/Python/3.10/lib/python/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>YADON_RESULT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M1I</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1L</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1V</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         YADON_RESULT\n",
       "MUTATION             \n",
       "M1I                 R\n",
       "M1L                 R\n",
       "M1V                 S"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "YADON = pandas.read_excel('./data/source-data/41467_2017_721_MOESM4_ESM.xlsx')\n",
    "\n",
    "# rename the columns to not contain spaces\n",
    "YADON.rename(columns={  'Substitution':'MUTATION',\n",
    "                        'Under-Represented?': 'UNDER_REPRESENTED',\n",
    "                        'Catalogue': 'CATALOGUE',\n",
    "                        'Miotto et. al.': 'MIOTTO2014',\n",
    "                        'Walker et. al.': 'WALKER2015'}, inplace=True)\n",
    "\n",
    "# make a Boolean IS_NONSYN columns\n",
    "def define_nonsyn(row):\n",
    "    IS_NONSYN = False\n",
    "    IS_SYN = False\n",
    "    CONTAINS_STOP = False\n",
    "    if row['MUTATION'][-1]=='*':\n",
    "        CONTAINS_STOP=True\n",
    "    if row['MUTATION'][0]==row['MUTATION'][-1]:\n",
    "        IS_SYN=True\n",
    "    elif row['MUTATION'][0]!=row['MUTATION'][-1]:\n",
    "        IS_NONSYN=True\n",
    "    return pandas.Series([IS_NONSYN,IS_SYN,CONTAINS_STOP])\n",
    "\n",
    "YADON[['IS_NONSYNONYMOUS','IS_SYNONYMOUS','CONTAINS_STOP']] = YADON.apply(define_nonsyn, axis=1)       \n",
    "YADON['UNDER_REPRESENTED'] = YADON.UNDER_REPRESENTED=='Y'\n",
    "\n",
    "# replace the NaNs with zeros and then convert to True/False                        \n",
    "for i in ['in vitro Resistant', 'in vitro Susceptible', 'in vivo Resistant', 'in vivo Susceptible','MIOTTO2014','WALKER2015']:\n",
    "    YADON[i] = YADON[i].fillna(0).astype('bool')                        \n",
    "\n",
    "# combine the one-hot encoded columns\n",
    "def assign_phenotypes(row, string):\n",
    "    if row[string+' Resistant'] and not row[string+' Susceptible']:\n",
    "        return('R')\n",
    "    elif not row[string+' Resistant'] and row[string+' Susceptible']:\n",
    "        return('S')\n",
    "    else:\n",
    "        return(None)\n",
    "\n",
    "YADON['IN_VITRO_PHENOTYPE'] = YADON.apply(assign_phenotypes, args=('in vitro',), axis=1)\n",
    "YADON['IN_VIVO_PHENOTYPE'] = YADON.apply(assign_phenotypes, args=('in vivo',), axis=1)\n",
    "\n",
    "# drop the one-hot columns\n",
    "YADON.drop(columns=['in vitro Resistant', 'in vitro Susceptible', 'in vivo Resistant', 'in vivo Susceptible'], inplace=True)\n",
    "\n",
    "def interpet_catalogue(row):\n",
    "    if row.CATALOGUE=='E':\n",
    "        return('R')\n",
    "    elif row.CATALOGUE=='D':\n",
    "        return('S')\n",
    "\n",
    "YADON['YADON_RESULT'] = YADON.apply(interpet_catalogue, axis=1)       \n",
    "\n",
    "def replace_stop_codon(row):\n",
    "    if \"*\" in row.MUTATION:\n",
    "        return row.MUTATION.replace(\"*\",\"!\")\n",
    "    else:\n",
    "        return row.MUTATION\n",
    "\n",
    "YADON['MUTATION'] = YADON.apply(replace_stop_codon, axis=1)\n",
    "\n",
    "# reorder the dataset\n",
    "YADON = YADON[['MUTATION', 'UNDER_REPRESENTED', 'IN_VITRO_PHENOTYPE', 'IN_VIVO_PHENOTYPE', 'YADON_RESULT', 'MIOTTO2014',\n",
    "       'WALKER2015', 'IS_NONSYNONYMOUS','IS_SYNONYMOUS','CONTAINS_STOP']]\n",
    "\n",
    "YADON = YADON[(~YADON.UNDER_REPRESENTED) & (YADON.YADON_RESULT.notna())]\n",
    "YADON.set_index('MUTATION', inplace=True, verify_integrity=True)\n",
    "YADON[['YADON_RESULT']].to_csv('data/catalogues/ds-yadon2017.csv', index=True)\n",
    "YADON[['YADON_RESULT']][:3]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1 (b). NEJM2018"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PREDICTION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-3_indel</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145_indel</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185_indel</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          PREDICTION\n",
       "MUTATION            \n",
       "-3_indel           S\n",
       "145_indel          S\n",
       "185_indel          R"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nejm2018 = pandas.read_csv('data/source-data/NC_000962.3_NEJM2018_v1.1_GARC1_RUS.csv')\n",
    "\n",
    "nejm2018 = nejm2018[(nejm2018.DRUG=='PZA') & (nejm2018.PREDICTION.isin(['R','S'])) & (~nejm2018.MUTATION.str.contains('*', regex=False))]\n",
    "\n",
    "def infer_type(row):\n",
    "    mut = row.MUTATION\n",
    "    cols = mut.split('@')\n",
    "    IS_SNP = True\n",
    "    IN_CDS = True\n",
    "    if 'ins' in mut:\n",
    "        IS_SNP=False\n",
    "    elif 'del' in mut:\n",
    "        IS_SNP=False\n",
    "    elif '-' in mut:\n",
    "        IN_CDS=False\n",
    "    return pandas.Series([IS_SNP, IN_CDS, cols[0], cols[1]])        \n",
    "\n",
    "nejm2018[['IS_SNP', 'IN_CDS', 'GENE', 'MUTATION']] = nejm2018.apply(infer_type, axis=1)    \n",
    "nejm2018.set_index('MUTATION', inplace=True, verify_integrity=True)\n",
    "nejm2018[['PREDICTION']].to_csv('data/catalogues/ds-nejm2018.csv', index=True)\n",
    "nejm2018[['PREDICTION']][:3]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1 (c). WHO2021"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R    38\n",
      "Name: PREDICTION, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PREDICTION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-5_indel</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108_indel</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116_indel</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          PREDICTION\n",
       "MUTATION            \n",
       "-5_indel           R\n",
       "108_indel          R\n",
       "116_indel          R"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "who = pandas.read_csv('data/source-data/NC_000962.3_WHO-UCN-GTB-PCI-2021.7_v1.0_GARC1_RUS.csv')\n",
    "\n",
    "who = who[(who.DRUG=='PZA') & (who.PREDICTION.isin(['R','S'])) & (~who.MUTATION.str.contains('*', regex=False))]\n",
    "\n",
    "def infer_type(row):\n",
    "    mut = row.MUTATION\n",
    "    cols = mut.split('@')\n",
    "    gene = cols[0]\n",
    "    mutation = cols[1]\n",
    "    IS_SNP = True\n",
    "    IN_CDS = True\n",
    "    if ('ins' in mut) or ('del' in mut):\n",
    "        IS_SNP=False\n",
    "        bits = mutation.split('_')\n",
    "        mutation = bits[0]+\"_indel\"\n",
    "    elif '-' in mut:\n",
    "        IN_CDS=False\n",
    "    return pandas.Series([IS_SNP, IN_CDS, gene, mutation])        \n",
    "\n",
    "who[['IS_SNP', 'IN_CDS', 'GENE', 'MUTATION']] = who.apply(infer_type, axis=1)    \n",
    "\n",
    "# By using a generic `X_indel` description there are now several different indels at a specific position, however they all confer resistance according to the catalogue so we can drop the duplicate rows\n",
    "foo = who[~who.IS_SNP].MUTATION.value_counts()\n",
    "print(who[who.MUTATION.isin(foo[foo>1].index)].PREDICTION.value_counts())\n",
    "\n",
    "who.set_index('MUTATION', inplace=True, verify_integrity=False)\n",
    "who = who.loc[~who.index.duplicated(keep='first')]\n",
    "who.reset_index(inplace=True)\n",
    "who.set_index('MUTATION', inplace=True, verify_integrity=True)\n",
    "who[['PREDICTION']].to_csv('data/catalogues/ds-who2021.csv', index=True)\n",
    "who[['PREDICTION']][:3]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Validation dataset\n",
    "\n",
    "This is derived from three sets of clinical samples with DST data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2 (a). CRyPTIC2021"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 22842 samples in the CRyPTIC dataset with both genetics and a pncA phenotype\n"
     ]
    }
   ],
   "source": [
    "DST_CORE = pandas.read_csv('data/source-data/cryptic1-dst.csv')\n",
    "DST_CORE.set_index('UNIQUEID', inplace=True)\n",
    "\n",
    "LINEAGES_CORE = pandas.read_csv('data/source-data/cryptic1-lineages.csv')\n",
    "LINEAGES_CORE.set_index('UNIQUEID', inplace=True)\n",
    "\n",
    "MUTATIONS_CORE = pandas.read_csv('data/source-data/cryptic1-mutations.csv')\n",
    "MUTATIONS_CORE.set_index('UNIQUEID', inplace=True)\n",
    "MUTATIONS_CORE = MUTATIONS_CORE.join(LINEAGES_CORE, how='left')\n",
    "\n",
    "print(\"There are %i samples in the CRyPTIC dataset with both genetics and a pncA phenotype\" % len(DST_CORE))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If a sample has a mutation in *pncA*, it could have one, two or more than two mutations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    6622\n",
       "2     228\n",
       "3       1\n",
       "Name: N_MUTATIONS, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS_CORE.reset_index(inplace=True)\n",
    "MUTATIONS_CORE_COUNTS=MUTATIONS_CORE[['UNIQUEID','SITEID']].groupby('UNIQUEID').count()\n",
    "MUTATIONS_CORE.set_index('UNIQUEID',inplace=True)\n",
    "MUTATIONS_CORE_COUNTS.rename(columns={'SITEID':'N_MUTATIONS'},inplace=True)\n",
    "MUTATIONS_CORE_COUNTS\n",
    "MUTATIONS_CORE_COUNTS.N_MUTATIONS.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some of these are likely to be lineage-associated. Let's check if any of the synoymous mutations are potential lineage-defining/associated mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>MYKROBE_LINEAGE_NAME_1</th>\n",
       "      <th>Lineage 3</th>\n",
       "      <th>Lineage 4</th>\n",
       "      <th>Lineage 5</th>\n",
       "      <th>Lineage Bovis</th>\n",
       "      <th>Mixed</th>\n",
       "      <th>Unknown</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A38A</th>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G75G</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G97G</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L117L</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L35L</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S65S</th>\n",
       "      <td>2872</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S67S</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S74S</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T100T</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "MYKROBE_LINEAGE_NAME_1  Lineage 3  Lineage 4  Lineage 5  Lineage Bovis  Mixed  \\\n",
       "MUTATION                                                                        \n",
       "A38A                            0         17          0              0      0   \n",
       "G75G                            0          5          0              0      0   \n",
       "G97G                            0          1          2              0      0   \n",
       "L117L                           3          0          0              0      0   \n",
       "L35L                            1          1          0              0      0   \n",
       "S65S                         2872          0          0              0     21   \n",
       "S67S                            2          0          0              0      0   \n",
       "S74S                            0          0          0              1      0   \n",
       "T100T                           2          0          0              0      0   \n",
       "\n",
       "MYKROBE_LINEAGE_NAME_1  Unknown  \n",
       "MUTATION                         \n",
       "A38A                          0  \n",
       "G75G                          0  \n",
       "G97G                          0  \n",
       "L117L                         0  \n",
       "L35L                          0  \n",
       "S65S                          1  \n",
       "S67S                          0  \n",
       "S74S                          0  \n",
       "T100T                         0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutation_counts = MUTATIONS_CORE.MUTATION.value_counts()\n",
    "\n",
    "df = MUTATIONS_CORE[MUTATIONS_CORE.MUTATION.isin(mutation_counts[mutation_counts>=3].index) & MUTATIONS_CORE.IS_SYNONYMOUS]\n",
    "\n",
    "pandas.crosstab(df.MUTATION, df.MYKROBE_LINEAGE_NAME_1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Synonymous mutations S65S and A38A are entirely associated with different lineages and have n>10 so we will assume are lineage-associated and so can be ignored. This reduces the number of samples with >=1 mutation present."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    3578\n",
       "2      33\n",
       "Name: N_MUTATIONS, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS_CORE = MUTATIONS_CORE[~MUTATIONS_CORE.MUTATION.isin(['S65S','A38A'])]\n",
    "MUTATIONS_CORE.reset_index(inplace=True)\n",
    "MUTATIONS_CORE_COUNTS=MUTATIONS_CORE[['UNIQUEID','SITEID']].groupby('UNIQUEID').count()\n",
    "MUTATIONS_CORE.set_index('UNIQUEID',inplace=True)\n",
    "MUTATIONS_CORE_COUNTS.rename(columns={'SITEID':'N_MUTATIONS'},inplace=True)\n",
    "MUTATIONS_CORE_COUNTS\n",
    "MUTATIONS_CORE_COUNTS.N_MUTATIONS.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overall we end up with 3,611 samples with DST data and one or more mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    19231\n",
       "True      3611\n",
       "Name: HAS_MUTATION, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_with_mutations=set(MUTATIONS_CORE.index)\n",
    "DST_CORE.loc[DST_CORE.index.isin(samples_with_mutations), 'HAS_MUTATION']=True\n",
    "DST_CORE.HAS_MUTATION.fillna(False,inplace=True)\n",
    "DST_CORE.HAS_MUTATION.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HAS_MUTATION</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>684</td>\n",
       "      <td>18547</td>\n",
       "      <td>19231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>2667</td>\n",
       "      <td>944</td>\n",
       "      <td>3611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>3351</td>\n",
       "      <td>19491</td>\n",
       "      <td>22842</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PHENOTYPE        R      S    All\n",
       "HAS_MUTATION                    \n",
       "False          684  18547  19231\n",
       "True          2667    944   3611\n",
       "All           3351  19491  22842"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pandas.crosstab(DST_CORE.HAS_MUTATION, DST_CORE.PHENOTYPE, margins=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "73.9 % of samples with a mutation are resistant, whilst only 3.6 % of samples without a mutation are resistant\n"
     ]
    }
   ],
   "source": [
    "prop_R_mut = 100*df['R'][True] / df['All'][True]\n",
    "prop_R_nomut = 100*df['R'][False] / df['All'][False]\n",
    "print(\"%.1f %% of samples with a mutation are resistant, whilst only %.1f %% of samples without a mutation are resistant\" % (prop_R_mut, prop_R_nomut))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's take a quick look at the most common mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "H57D     242\n",
       "a-11g    134\n",
       "I6L      128\n",
       "Q10R      98\n",
       "Q10P      82\n",
       "H51R      79\n",
       "G97D      58\n",
       "L4S       47\n",
       "H57R      46\n",
       "V139A     46\n",
       "Name: MUTATION, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS_CORE.MUTATION.value_counts()[:10]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we need to identify those samples which have one (and only one) mutation in *pncA*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>IN_PROMOTER</th>\n",
       "      <th>IS_SYNONYMOUS</th>\n",
       "      <th>IS_NONSYNONYMOUS</th>\n",
       "      <th>IS_HET</th>\n",
       "      <th>IS_NULL</th>\n",
       "      <th>IS_FILTER_PASS</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "      <th>MYKROBE_LINEAGE_NAME_1</th>\n",
       "      <th>MYKROBE_LINEAGE_NAME_2</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>SOURCE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.00.subj.1000347.lab.H111540004.iso.1</th>\n",
       "      <td>pncA</td>\n",
       "      <td>Q141P</td>\n",
       "      <td>141.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cag</td>\n",
       "      <td>ccg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>00</td>\n",
       "      <td>1</td>\n",
       "      <td>Lineage 2</td>\n",
       "      <td>lineage2.2.10</td>\n",
       "      <td>R</td>\n",
       "      <td>NEJM2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.00.subj.1000595.lab.H123460044.iso.1</th>\n",
       "      <td>pncA</td>\n",
       "      <td>D12A</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gac</td>\n",
       "      <td>gcc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>00</td>\n",
       "      <td>1</td>\n",
       "      <td>Lineage 2</td>\n",
       "      <td>lineage2.2.9</td>\n",
       "      <td>S</td>\n",
       "      <td>NEJM2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.00.subj.1004213.lab.H111060034.iso.1</th>\n",
       "      <td>pncA</td>\n",
       "      <td>392_indel</td>\n",
       "      <td>392.0</td>\n",
       "      <td>131.0</td>\n",
       "      <td>2288850.0</td>\n",
       "      <td>392.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>2.0</td>\n",
       "      <td>392_ins</td>\n",
       "      <td>392_ins_2</td>\n",
       "      <td>00</td>\n",
       "      <td>0</td>\n",
       "      <td>Lineage 2</td>\n",
       "      <td>lineage2.2.6</td>\n",
       "      <td>R</td>\n",
       "      <td>NEJM2018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           GENE   MUTATION  POSITION  \\\n",
       "UNIQUEID                                                               \n",
       "site.00.subj.1000347.lab.H111540004.iso.1  pncA      Q141P     141.0   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1  pncA       D12A      12.0   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1  pncA  392_indel     392.0   \n",
       "\n",
       "                                           AMINO_ACID_NUMBER  GENOME_INDEX  \\\n",
       "UNIQUEID                                                                     \n",
       "site.00.subj.1000347.lab.H111540004.iso.1              141.0           NaN   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1               12.0           NaN   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1              131.0     2288850.0   \n",
       "\n",
       "                                           NUCLEOTIDE_NUMBER  REF  ALT  \\\n",
       "UNIQUEID                                                                 \n",
       "site.00.subj.1000347.lab.H111540004.iso.1                NaN  cag  ccg   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1                NaN  gac  gcc   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1              392.0  NaN  NaN   \n",
       "\n",
       "                                           IS_SNP  IS_INDEL  IN_CDS  \\\n",
       "UNIQUEID                                                              \n",
       "site.00.subj.1000347.lab.H111540004.iso.1    True     False    True   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1    True     False    True   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1   False      True    True   \n",
       "\n",
       "                                           IN_PROMOTER  IS_SYNONYMOUS  \\\n",
       "UNIQUEID                                                                \n",
       "site.00.subj.1000347.lab.H111540004.iso.1        False          False   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1        False          False   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1        False          False   \n",
       "\n",
       "                                           IS_NONSYNONYMOUS  IS_HET  IS_NULL  \\\n",
       "UNIQUEID                                                                       \n",
       "site.00.subj.1000347.lab.H111540004.iso.1              True   False    False   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1              True   False    False   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1             False   False    False   \n",
       "\n",
       "                                           IS_FILTER_PASS ELEMENT_TYPE  \\\n",
       "UNIQUEID                                                                 \n",
       "site.00.subj.1000347.lab.H111540004.iso.1            True         GENE   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1            True         GENE   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1            True         GENE   \n",
       "\n",
       "                                          MUTATION_TYPE  INDEL_LENGTH  \\\n",
       "UNIQUEID                                                                \n",
       "site.00.subj.1000347.lab.H111540004.iso.1           AAM           NaN   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1           AAM           NaN   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1         INDEL           2.0   \n",
       "\n",
       "                                           INDEL_1    INDEL_2 SITEID  \\\n",
       "UNIQUEID                                                               \n",
       "site.00.subj.1000347.lab.H111540004.iso.1      NaN        NaN     00   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1      NaN        NaN     00   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1  392_ins  392_ins_2     00   \n",
       "\n",
       "                                           NUMBER_NUCLEOTIDE_CHANGES  \\\n",
       "UNIQUEID                                                               \n",
       "site.00.subj.1000347.lab.H111540004.iso.1                          1   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1                          1   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1                          0   \n",
       "\n",
       "                                          MYKROBE_LINEAGE_NAME_1  \\\n",
       "UNIQUEID                                                           \n",
       "site.00.subj.1000347.lab.H111540004.iso.1              Lineage 2   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1              Lineage 2   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1              Lineage 2   \n",
       "\n",
       "                                          MYKROBE_LINEAGE_NAME_2 PHENOTYPE  \\\n",
       "UNIQUEID                                                                     \n",
       "site.00.subj.1000347.lab.H111540004.iso.1          lineage2.2.10         R   \n",
       "site.00.subj.1000595.lab.H123460044.iso.1           lineage2.2.9         S   \n",
       "site.00.subj.1004213.lab.H111060034.iso.1           lineage2.2.6         R   \n",
       "\n",
       "                                             SOURCE  \n",
       "UNIQUEID                                             \n",
       "site.00.subj.1000347.lab.H111540004.iso.1  NEJM2018  \n",
       "site.00.subj.1000595.lab.H123460044.iso.1  NEJM2018  \n",
       "site.00.subj.1004213.lab.H111060034.iso.1  NEJM2018  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DST_CORE=DST_CORE.join(MUTATIONS_CORE_COUNTS,how='left')\n",
    "DST_CORE.N_MUTATIONS.fillna(0,inplace=True) \n",
    "DST_CORE['IS_SOLO']=DST_CORE.N_MUTATIONS==1\n",
    "SOLOS=MUTATIONS_CORE.join(DST_CORE[DST_CORE.IS_SOLO][['PHENOTYPE', 'SOURCE']],how='inner')\n",
    "SOLOS[:3]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can summarise and write to disc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-29_indel</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2_indel</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-32_indel</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-3_indel</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-4_indel</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g-9a</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t-10c</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t-12c</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t-7c</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t-7g</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>561 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "PHENOTYPE                 R   S\n",
       "MUTATION  IS_SNP IN_CDS        \n",
       "-29_indel False  False    0   1\n",
       "-2_indel  False  False    0  12\n",
       "-32_indel False  False    0   1\n",
       "-3_indel  False  False    2   0\n",
       "-4_indel  False  False   10   1\n",
       "...                      ..  ..\n",
       "g-9a      True   False    0   1\n",
       "t-10c     True   False    0   2\n",
       "t-12c     True   False    9   1\n",
       "t-7c      True   False    5   0\n",
       "t-7g      True   False    1   0\n",
       "\n",
       "[561 rows x 2 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SOLO_MUTATIONS_CROSSTAB=pandas.crosstab([SOLOS.MUTATION,SOLOS.IS_SNP,SOLOS.IN_CDS],SOLOS.PHENOTYPE,margins=False,margins_name='TOTAL')\n",
    "SOLO_MUTATIONS_CROSSTAB.to_csv('data/clinical-samples/ds-cryptic2021.csv')\n",
    "SOLO_MUTATIONS_CROSSTAB"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2 (b). Miotto2014 data\n",
    "\n",
    "> Miotto, P., Cabibbe, A. M., Feuerriegel, S., Casali, N., Drobniewski, F., Rodionova, Y., Bakonyte, D., Stakenas, P., Pimkina, E., Augustynowicz-Kopeć, E., Degano, M., Ambrosi, A., Hoffner, S., Mansjö, M., Werngren, J., Rüsch-Gerdes, S., Niemann, S., & Cirillo, D. M. (2014). Mycobacterium tuberculosis pyrazinamide resistance determinants: a multicenter study. mBio, 5(5), e01819-14. https://doi.org/10.1128/mBio.01819-14\n",
    "\n",
    "Load the raw Excel file from the supplement of this paper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-3_indel</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-5_indel</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A102P</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A102R</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A102T</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>a-11g</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>31</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>a-11t</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g-13t</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t-7c</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t-7g</th>\n",
       "      <th>True</th>\n",
       "      <th>False</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>199 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "PHENOTYPE                R  S\n",
       "MUTATION IS_SNP IN_CDS       \n",
       "-3_indel False  False    1  1\n",
       "-5_indel False  False    2  0\n",
       "A102P    True   True     0  1\n",
       "A102R    True   True     0  1\n",
       "A102T    True   True     1  0\n",
       "...                     .. ..\n",
       "a-11g    True   False   31  4\n",
       "a-11t    True   False    0  1\n",
       "g-13t    True   False    0  1\n",
       "t-7c     True   False    4  0\n",
       "t-7g     True   False    1  0\n",
       "\n",
       "[199 rows x 2 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MIOTTO2014 = pandas.read_excel('./data/source-data/mbo005142032st1.xlsx', header=2)\n",
    "\n",
    "def lookup_mutation(row):\n",
    "\n",
    "    mutation = None\n",
    "    IS_SNP = None\n",
    "    IN_CDS = None\n",
    "    \n",
    "    # skipping rows with a comma skips samples with multiple mutations\n",
    "    if row['pncA aa'] != 'WT' and ',' not in row['pncA nt'] and ',' not in row['pncA aa'] and row['pncA nt'] != 'del >200 nt':\n",
    "        if row['pncA aa'][:3] not in amino_acid_lookup.keys():\n",
    "            if 'promoter' in row['pncA aa']: # or 'shift' not in row['pncA aa']:\n",
    "                IN_CDS = False\n",
    "                if row['pncA nt'][:3]  in ['Ins', 'Del']: \n",
    "                    IS_SNP = False\n",
    "                    mut = row['pncA nt']\n",
    "                    pos = mut.split(' ')[0][3:]\n",
    "                    mutation = (str(int(pos))+'_indel')\n",
    "                else:\n",
    "                    IS_SNP = True\n",
    "                    mut = row['pncA nt']\n",
    "                    assert mut[0] in ['a','t','c','g'], mut\n",
    "                    assert mut[-1] in ['a','t','c','g'], mut\n",
    "                    mutation = (row['pncA nt'])\n",
    "        else:\n",
    "            mut = row['pncA aa']\n",
    "            IS_SNP = True\n",
    "            IN_CDS = True\n",
    "            if 'stop' in mut:\n",
    "                if mut[:4] == 'stop':\n",
    "                    mutation = (amino_acid_lookup[mut[:4]] + mut[4:-3] + amino_acid_lookup[mut[-3:]])\n",
    "                else:\n",
    "                    mutation = (amino_acid_lookup[mut[:3]] + mut[3:-4] + amino_acid_lookup[mut[-4:]])\n",
    "            else:\n",
    "                ref = amino_acid_lookup[mut[:3]]\n",
    "                pos = mut[3:-3]\n",
    "                alt = amino_acid_lookup[mut[-3:]]\n",
    "                assert pncA_amino_acid_sequence[int(pos)-1] == ref, mut\n",
    "\n",
    "                mutation = (ref + pos + alt)\n",
    "    return pandas.Series([mutation, IS_SNP, IN_CDS])\n",
    "    \n",
    "MIOTTO2014[['MUTATION', 'IS_SNP', 'IN_CDS']] = MIOTTO2014.apply(lookup_mutation, axis=1)\n",
    "MIOTTO2014.rename(columns = {'PZA': 'PHENOTYPE'}, inplace=True)\n",
    "MIOTTO2014_MUTATIONS_SUMMARY=pandas.crosstab([MIOTTO2014.MUTATION,MIOTTO2014.IS_SNP,MIOTTO2014.IN_CDS],MIOTTO2014.PHENOTYPE,margins=False)\n",
    "MIOTTO2014_MUTATIONS_SUMMARY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are a total of 755 samples in the Miotto2014 dataset\n"
     ]
    }
   ],
   "source": [
    "n_miotto = MIOTTO2014_MUTATIONS_SUMMARY.R.sum() + MIOTTO2014_MUTATIONS_SUMMARY.S.sum()\n",
    "print(\"There are a total of \" + str(n_miotto) + \" samples in the Miotto2014 dataset\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 191 non-synoymous mutations in 704 samples\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-3_indel</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-5_indel</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A102P</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PHENOTYPE               R  S\n",
       "MUTATION IS_SNP IN_CDS      \n",
       "-3_indel False  False   1  1\n",
       "-5_indel False  False   2  0\n",
       "A102P    True   True    0  1"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MIOTTO2014_MUTATIONS_SUMMARY.reset_index(inplace=True)\n",
    "n_miotto_nonsyn = MIOTTO2014_MUTATIONS_SUMMARY[(MIOTTO2014_MUTATIONS_SUMMARY.IS_SNP) & (MIOTTO2014_MUTATIONS_SUMMARY.IN_CDS)]\n",
    "print(\"There are \" + str(len(n_miotto_nonsyn)) + \" non-synoymous mutations in \" + str(n_miotto_nonsyn.R.sum() + n_miotto_nonsyn.S.sum()) + \" samples\")\n",
    "MIOTTO2014_MUTATIONS_SUMMARY.set_index(['MUTATION', 'IS_SNP', 'IN_CDS'], inplace=True)\n",
    "MIOTTO2014_MUTATIONS_SUMMARY.to_csv('data/clinical-samples/ds-miotto2014.csv',index=True)\n",
    "MIOTTO2014_MUTATIONS_SUMMARY[:3]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2 (c). Whitfield2015\n",
    "\n",
    "> Whitfield, M. G., Soeters, H. M., Warren, R. M., York, T., Sampson, S. L., Streicher, E. M., Van Helden, P. D., & Van Rie, A. (2015). A global perspective on pyrazinamide resistance: Systematic review and meta-analysis. PLoS ONE, 10(7), 1–16. https://doi.org/10.1371/journal.pone.0133869\n",
    "\n",
    "Load in the Whitfield data -- this was copied from `zjm999094564so1.pdf`, missing out any promoter/indels, into an Excel sheet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 65 non-synonymous mutations in 634 samples in the Whitfield2015 dataset.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>I5T</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I6L</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V9V</th>\n",
       "      <th>True</th>\n",
       "      <th>True</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        R    S\n",
       "MUTATION IS_SNP IN_CDS        \n",
       "I5T      True   True    0    1\n",
       "I6L      True   True    0  128\n",
       "V9V      True   True    0    1"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "WHITFIELD2015 = pandas.read_excel('./data/source-data/zjm999094564so1.xlsx', header=1)\n",
    "\n",
    "def parse_whitfield2015(row):\n",
    "\n",
    "    mut = row.MUTATION.rstrip()\n",
    "    IS_SNP = True\n",
    "    IN_CDS = True\n",
    "\n",
    "    if mut[:4]== 'stop':\n",
    "        ref = amino_acid_lookup[mut[:4]]\n",
    "        pos = mut[4:-3]\n",
    "        alt = amino_acid_lookup[mut[-3:]]\n",
    "    elif mut[-4:] == 'stop':\n",
    "        ref = amino_acid_lookup[mut[:3]]\n",
    "        pos = mut[3:-4]\n",
    "        alt = amino_acid_lookup[mut[-4:]]\n",
    "    else:\n",
    "        ref = amino_acid_lookup[mut[:3]]\n",
    "        pos = mut[3:-3]\n",
    "        alt = amino_acid_lookup[mut[-3:]]\n",
    "\n",
    "    mutation = ref + pos + alt\n",
    "    if int(pos) <= 186:\n",
    "        assert pncA_amino_acid_sequence[int(pos)-1] == ref, mut\n",
    "    else:\n",
    "        assert '!' == ref , mut\n",
    "\n",
    "    \n",
    "    return(pandas.Series([mutation, IS_SNP, IN_CDS]))\n",
    "\n",
    "WHITFIELD2015[['MUTATION', 'IS_SNP', 'IN_CDS']] = WHITFIELD2015.apply(parse_whitfield2015, axis=1)\n",
    "WHITFIELD2015 = WHITFIELD2015[WHITFIELD2015.MUTATION.notna()][['MUTATION', 'IS_SNP', 'IN_CDS', 'R', 'S']]\n",
    "WHITFIELD2015.set_index(['MUTATION', 'IS_SNP', 'IN_CDS'], inplace=True, verify_integrity=True)\n",
    "WHITFIELD2015.to_csv('data/clinical-samples/ds-whitfield2015.csv')\n",
    "print(\"There are \" + str(len(WHITFIELD2015)) + \" non-synonymous mutations in \" + str(WHITFIELD2015.R.sum() + WHITFIELD2015.S.sum()) + \" samples in the Whitfield2015 dataset.\")\n",
    "WHITFIELD2015[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "7a792fcb311f9eb9f3c1b942a8c87ada8484712b89b670347c16a1088e0a1f69"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
