{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sbmlcore, pandas, numpy, copy\n",
    "import itertools\n",
    "\n",
    "pandas.options.display.max_columns=999\n",
    "pandas.options.display.max_rows=150"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in three sets of clinical samples with mutations and recorded phenotypes to form the VALIDATION dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of mutations and samples in each set are:\n",
      "miotto2014 199 755\n",
      "whitfield2015 65 634\n",
      "cryptic2021 561 3578\n",
      "\n",
      "..and the intersection in numbers of mutations between sets are:\n",
      "('miotto2014', 'whitfield2015') 41\n",
      "('miotto2014', 'cryptic2021') 179\n",
      "('whitfield2015', 'cryptic2021') 47\n",
      "\n",
      "..and the number of mutations in all three sets is:\n",
      "('miotto2014', 'whitfield2015', 'cryptic2021') 35\n"
     ]
    }
   ],
   "source": [
    "filestem = 'data/ds-validation'\n",
    "\n",
    "validation_csvs = ['miotto2014','whitfield2015','cryptic2021']\n",
    "stem='data/clinical-samples/ds-'\n",
    "validation_dfs = {}\n",
    "validation_sets = {}\n",
    "for i in validation_csvs:\n",
    "    validation_dfs[i] = pandas.read_csv(stem+i+'.csv')\n",
    "    validation_dfs[i].set_index('MUTATION', inplace=True)\n",
    "    validation_sets[i] = set(validation_dfs[i].index)\n",
    "    validation_dfs[i].reset_index(inplace=True)\n",
    "    validation_dfs[i].set_index(['MUTATION', 'IS_SNP', 'IN_CDS'], inplace=True)\n",
    "    validation_dfs[i]['total'] = validation_dfs[i].R + validation_dfs[i].S\n",
    "\n",
    "print(\"The number of mutations and samples in each set are:\")\n",
    "for i in validation_csvs:\n",
    "    print(i, len(validation_sets[i]), validation_dfs[i].total.sum())\n",
    "\n",
    "print(\"\\n..and the intersection in numbers of mutations between sets are:\")\n",
    "if len(validation_csvs)>1:\n",
    "    for i in list(itertools.combinations(validation_csvs,2)):\n",
    "        print(i, len(validation_sets[i[0]].intersection(validation_sets[i[1]])))\n",
    "\n",
    "print(\"\\n..and the number of mutations in all three sets is:\")\n",
    "if len(validation_csvs)>2:\n",
    "    for i in list(itertools.combinations(validation_csvs,3)):\n",
    "        print(i, len(validation_sets[i[0]] & validation_sets[i[1]] & validation_sets[i[2]]))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's join them all and aggregate the AST results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>index</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>TOTAL</th>\n",
       "      <th>PROP_R</th>\n",
       "      <th>PROP_S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!187G</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!187R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-29_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-2_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-32_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588</th>\n",
       "      <td>g-9a</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>589</th>\n",
       "      <td>t-10c</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>590</th>\n",
       "      <td>t-12c</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591</th>\n",
       "      <td>t-7c</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>t-7g</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>593 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "index   MUTATION  IS_SNP  IN_CDS  R   S  TOTAL  PROP_R  PROP_S\n",
       "0          !187G    True    True  1   1      2     0.5     0.5\n",
       "1          !187R    True    True  0   1      1     0.0     1.0\n",
       "2      -29_indel   False   False  0   1      1     0.0     1.0\n",
       "3       -2_indel   False   False  0  12     12     0.0     1.0\n",
       "4      -32_indel   False   False  0   1      1     0.0     1.0\n",
       "..           ...     ...     ... ..  ..    ...     ...     ...\n",
       "588         g-9a    True   False  0   1      1     0.0     1.0\n",
       "589        t-10c    True   False  0   2      2     0.0     1.0\n",
       "590        t-12c    True   False  9   1     10     0.9     0.1\n",
       "591         t-7c    True   False  9   0      9     1.0     0.0\n",
       "592         t-7g    True   False  2   0      2     1.0     0.0\n",
       "\n",
       "[593 rows x 8 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VALIDATION = validation_dfs['cryptic2021'].join(validation_dfs['miotto2014'], lsuffix='s', rsuffix = 'm', how='outer')\n",
    "VALIDATION = VALIDATION.join(validation_dfs['whitfield2015'], rsuffix='w',how='outer')\n",
    "VALIDATION.rename(columns={'R': 'Rw', 'S': 'Sw', 'total': 'totalw'}, inplace=True)\n",
    "VALIDATION.fillna(0, inplace=True)   \n",
    "VALIDATION = VALIDATION.astype('int')\n",
    "VALIDATION['R'] = VALIDATION['Rs'] + VALIDATION['Rm'] + VALIDATION['Rw']\n",
    "VALIDATION['S'] = VALIDATION['Ss'] + VALIDATION['Sm'] + VALIDATION['Sw']\n",
    "VALIDATION['TOTAL'] = VALIDATION['R'] + VALIDATION['S']\n",
    "VALIDATION.drop(columns=['Rs', 'Ss', 'Rm', 'Sm', 'Rw', 'Sw', 'totals', 'totalm', 'totalw'], inplace=True)\n",
    "VALIDATION.reset_index(inplace=True)\n",
    "VALIDATION['PROP_R']=VALIDATION['R']/VALIDATION['TOTAL']\n",
    "VALIDATION['PROP_S']=VALIDATION['S']/VALIDATION['TOTAL']\n",
    "VALIDATION.columns.name='index'\n",
    "VALIDATION"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Apply the arbitrary rules described in the Methods to assign an overall phenotype "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "307 samples cannot be assigned a phenotype\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "R    221\n",
       "S     65\n",
       "Name: PHENOTYPE, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def reliable_phenotype(row):\n",
    "    reliable_phenotype=False\n",
    "    phenotype='U'\n",
    "    if row['TOTAL']>=4:\n",
    "        if row['PROP_R']>=0.75:\n",
    "            reliable_phenotype=True\n",
    "            phenotype='R'\n",
    "        elif row['PROP_S']>=0.75:\n",
    "            reliable_phenotype=True\n",
    "            phenotype='S'\n",
    "    elif row['TOTAL']>=2:\n",
    "        if row['R']==row['TOTAL']:\n",
    "            reliable_phenotype=True\n",
    "            phenotype='R'\n",
    "        elif row['S']==row['TOTAL']:\n",
    "            reliable_phenotype=True\n",
    "            phenotype='S'\n",
    "\n",
    "    return pandas.Series([reliable_phenotype, phenotype]) \n",
    "\n",
    "VALIDATION[['RELIABLE_PHENOTYPE', 'PHENOTYPE']]=VALIDATION.apply(reliable_phenotype,axis=1)   \n",
    "\n",
    "print(\"%i samples cannot be assigned a phenotype\" % ((~VALIDATION.RELIABLE_PHENOTYPE).sum()))\n",
    "\n",
    "VALIDATION[VALIDATION.RELIABLE_PHENOTYPE].PHENOTYPE.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's take a look at the mutations that cannot be assigned a phenotype using these rules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>index</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>TOTAL</th>\n",
       "      <th>PROP_R</th>\n",
       "      <th>PROP_S</th>\n",
       "      <th>RELIABLE_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!187G</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!187R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-29_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-32_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-9_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>102_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>108_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>109_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>116_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>117_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "index   MUTATION  IS_SNP  IN_CDS  R  S  TOTAL  PROP_R  PROP_S  \\\n",
       "0          !187G    True    True  1  1      2     0.5     0.5   \n",
       "1          !187R    True    True  0  1      1     0.0     1.0   \n",
       "2      -29_indel   False   False  0  1      1     0.0     1.0   \n",
       "4      -32_indel   False   False  0  1      1     0.0     1.0   \n",
       "8       -9_indel   False   False  0  1      1     0.0     1.0   \n",
       "9      102_indel   False    True  1  0      1     1.0     0.0   \n",
       "10     108_indel   False    True  1  0      1     1.0     0.0   \n",
       "11     109_indel   False    True  1  0      1     1.0     0.0   \n",
       "12     116_indel   False    True  1  0      1     1.0     0.0   \n",
       "13     117_indel   False    True  1  0      1     1.0     0.0   \n",
       "\n",
       "index  RELIABLE_PHENOTYPE PHENOTYPE  \n",
       "0                   False         U  \n",
       "1                   False         U  \n",
       "2                   False         U  \n",
       "4                   False         U  \n",
       "8                   False         U  \n",
       "9                   False         U  \n",
       "10                  False         U  \n",
       "11                  False         U  \n",
       "12                  False         U  \n",
       "13                  False         U  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VALIDATION[~VALIDATION.RELIABLE_PHENOTYPE][:10]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Many of them only have a single measurement:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1     229\n",
       "2      14\n",
       "3      16\n",
       "4       3\n",
       "5       2\n",
       "6       8\n",
       "7       3\n",
       "8       3\n",
       "9       3\n",
       "10      2\n",
       "11      1\n",
       "12      2\n",
       "13      1\n",
       "14      1\n",
       "15      2\n",
       "17      1\n",
       "18      1\n",
       "19      1\n",
       "20      1\n",
       "22      1\n",
       "23      2\n",
       "24      1\n",
       "25      2\n",
       "27      1\n",
       "29      1\n",
       "51      1\n",
       "55      1\n",
       "59      1\n",
       "76      1\n",
       "84      1\n",
       "Name: TOTAL, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VALIDATION[~VALIDATION.RELIABLE_PHENOTYPE].TOTAL.value_counts().sort_index()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looking at the mutations with more than one sample we find an ambigious distribution of R and S results suggesting perhaps these are close to the MGIT breakpoint."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>index</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>TOTAL</th>\n",
       "      <th>PROP_R</th>\n",
       "      <th>PROP_S</th>\n",
       "      <th>RELIABLE_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!187G</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>201_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>257_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>477_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>48_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>67_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>A102P</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>A102V</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>25</td>\n",
       "      <td>0.480000</td>\n",
       "      <td>0.520000</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>A134G</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>A171E</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "index   MUTATION  IS_SNP  IN_CDS   R   S  TOTAL    PROP_R    PROP_S  \\\n",
       "0          !187G    True    True   1   1      2  0.500000  0.500000   \n",
       "34     201_indel   False    True   3   4      7  0.428571  0.571429   \n",
       "46     257_indel   False    True   1   1      2  0.500000  0.500000   \n",
       "112    477_indel   False    True   2   1      3  0.666667  0.333333   \n",
       "116     48_indel   False    True   1   1      2  0.500000  0.500000   \n",
       "136     67_indel   False    True   2   1      3  0.666667  0.333333   \n",
       "145        A102P    True    True   8   4     12  0.666667  0.333333   \n",
       "148        A102V    True    True  12  13     25  0.480000  0.520000   \n",
       "150        A134G    True    True   1   1      2  0.500000  0.500000   \n",
       "162        A171E    True    True   2   2      4  0.500000  0.500000   \n",
       "\n",
       "index  RELIABLE_PHENOTYPE PHENOTYPE  \n",
       "0                   False         U  \n",
       "34                  False         U  \n",
       "46                  False         U  \n",
       "112                 False         U  \n",
       "116                 False         U  \n",
       "136                 False         U  \n",
       "145                 False         U  \n",
       "148                 False         U  \n",
       "150                 False         U  \n",
       "162                 False         U  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VALIDATION[(~VALIDATION.RELIABLE_PHENOTYPE) & (VALIDATION.TOTAL>1)][:10]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's remove all these samples where a phenotype cannot be inferred using the arbitary rules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "VALIDATION = VALIDATION[VALIDATION.RELIABLE_PHENOTYPE]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_88397/858835743.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CLINICAL_SAMPLES[['IN_CDS', 'IS_SNP', 'IS_NONSYN', 'IS_MISSENSE']] = CLINICAL_SAMPLES.apply(classify_variant, axis=1)\n",
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_88397/858835743.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CLINICAL_SAMPLES[['IN_CDS', 'IS_SNP', 'IS_NONSYN', 'IS_MISSENSE']] = CLINICAL_SAMPLES.apply(classify_variant, axis=1)\n",
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_88397/858835743.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CLINICAL_SAMPLES[['IN_CDS', 'IS_SNP', 'IS_NONSYN', 'IS_MISSENSE']] = CLINICAL_SAMPLES.apply(classify_variant, axis=1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>False</th>\n",
       "      <th colspan=\"3\" halign=\"left\">True</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IS_NONSYN</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <th colspan=\"2\" halign=\"left\">True</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IS_MISSENSE</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <th>False</th>\n",
       "      <th>True</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IN_CDS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>45</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "IS_SNP      False True             \n",
       "IS_NONSYN   False False True       \n",
       "IS_MISSENSE False False False True \n",
       "IN_CDS                             \n",
       "False           4     8     0     0\n",
       "True           45    12    13   195"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def classify_variant(row):\n",
    "    is_cds=False\n",
    "    is_snp=False\n",
    "    is_nonsyn=False\n",
    "    is_missense=False\n",
    "    cols = row.MUTATION.split('_')\n",
    "    if '-' not in row.MUTATION:\n",
    "        is_cds=True        \n",
    "        if len(cols)!=3 and 'indel' not in row.MUTATION:\n",
    "            is_snp=True\n",
    "            if row.MUTATION[0]!=row.MUTATION[-1]:\n",
    "                is_nonsyn=True\n",
    "                if row.MUTATION[-1]!='!':\n",
    "                    is_missense=True\n",
    "    else:\n",
    "        if len(cols)!=3 and 'indel' not in row.MUTATION:\n",
    "            is_snp=True\n",
    "    \n",
    "    return pandas.Series([is_cds,is_snp,is_nonsyn,is_missense])    \n",
    "\n",
    "VALIDATION[['IN_CDS', 'IS_SNP', 'IS_NONSYN', 'IS_MISSENSE']] = VALIDATION.apply(classify_variant, axis=1)\n",
    "\n",
    "pandas.crosstab(VALIDATION.IN_CDS,[VALIDATION.IS_SNP, VALIDATION.IS_NONSYN, VALIDATION.IS_MISSENSE])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hence we find there are 12 promoter mutations (incl. 4 indels) and 45 indels in the CDS, with 12 synoymous mutations, 13 nonsense mutations and 195 missense mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1: There are 195 non-redundant missense mutations in this dataset\n"
     ]
    }
   ],
   "source": [
    "print(\"Table 1: There are %i non-redundant missense mutations in this dataset\" % (len(VALIDATION[ VALIDATION.IS_MISSENSE])))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "But we cannot structurally model (i) mutations in the Stop codon and also (ii) mutations in resid 186 since it is not resolved in the protein structure so let's identify these"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_88397/2247709330.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CLINICAL_SAMPLES['STRUCTURALLY_VALID'] = CLINICAL_SAMPLES.apply(valid_for_structure, axis=1)\n"
     ]
    }
   ],
   "source": [
    "def valid_for_structure(row):\n",
    "    if row.IN_CDS and row.IS_SNP and row.IS_NONSYN and row.IS_MISSENSE:\n",
    "        if \"!\" in row.MUTATION:\n",
    "            return False\n",
    "        elif '186' in row.MUTATION:\n",
    "            return False\n",
    "        else:\n",
    "            return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "VALIDATION['STRUCTURALLY_VALID'] = VALIDATION.apply(valid_for_structure, axis=1)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R    155\n",
       "S     40\n",
       "Name: PHENOTYPE, dtype: int64"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VALIDATION[ VALIDATION.IS_MISSENSE].PHENOTYPE.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "But how many of these can we not map onto the protein structure?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R    155\n",
       "S     40\n",
       "Name: PHENOTYPE, dtype: int64"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VALIDATION[(VALIDATION.STRUCTURALLY_VALID) & (VALIDATION.IS_MISSENSE)].PHENOTYPE.value_counts(dropna=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "None are lost for this dataset, but let's subset down anyway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET=VALIDATION[(VALIDATION.RELIABLE_PHENOTYPE) & (VALIDATION.STRUCTURALLY_VALID) & (VALIDATION.IS_MISSENSE)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>index</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>TOTAL</th>\n",
       "      <th>PROP_R</th>\n",
       "      <th>PROP_S</th>\n",
       "      <th>RELIABLE_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>IS_NONSYN</th>\n",
       "      <th>IS_MISSENSE</th>\n",
       "      <th>STRUCTURALLY_VALID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>A102R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>S</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>A102T</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>A134V</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>25</td>\n",
       "      <td>0.920000</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>A143G</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>A143T</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>True</td>\n",
       "      <td>S</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>W68S</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>Y103D</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>Y103H</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>Y103S</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>Y34D</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>0.846154</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>195 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "index MUTATION  IS_SNP  IN_CDS   R  S  TOTAL    PROP_R    PROP_S  \\\n",
       "146      A102R    True    True   0  3      3  0.000000  1.000000   \n",
       "147      A102T    True    True   3  0      3  1.000000  0.000000   \n",
       "151      A134V    True    True  23  2     25  0.920000  0.080000   \n",
       "152      A143G    True    True   6  0      6  1.000000  0.000000   \n",
       "154      A143T    True    True   1  6      7  0.142857  0.857143   \n",
       "..         ...     ...     ...  .. ..    ...       ...       ...   \n",
       "536       W68S    True    True   2  0      2  1.000000  0.000000   \n",
       "539      Y103D    True    True   4  0      4  1.000000  0.000000   \n",
       "540      Y103H    True    True   5  0      5  1.000000  0.000000   \n",
       "541      Y103S    True    True   2  0      2  1.000000  0.000000   \n",
       "545       Y34D    True    True  11  2     13  0.846154  0.153846   \n",
       "\n",
       "index  RELIABLE_PHENOTYPE PHENOTYPE  IS_NONSYN  IS_MISSENSE  \\\n",
       "146                  True         S       True         True   \n",
       "147                  True         R       True         True   \n",
       "151                  True         R       True         True   \n",
       "152                  True         R       True         True   \n",
       "154                  True         S       True         True   \n",
       "..                    ...       ...        ...          ...   \n",
       "536                  True         R       True         True   \n",
       "539                  True         R       True         True   \n",
       "540                  True         R       True         True   \n",
       "541                  True         R       True         True   \n",
       "545                  True         R       True         True   \n",
       "\n",
       "index  STRUCTURALLY_VALID  \n",
       "146                  True  \n",
       "147                  True  \n",
       "151                  True  \n",
       "152                  True  \n",
       "154                  True  \n",
       "..                    ...  \n",
       "536                  True  \n",
       "539                  True  \n",
       "540                  True  \n",
       "541                  True  \n",
       "545                  True  \n",
       "\n",
       "[195 rows x 13 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATASET\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_88397/953378862.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  DATASET.rename(columns={'PHENOTYPE':'CONSISTENT_PHENOTYPE'}, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>index</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>TOTAL</th>\n",
       "      <th>PROP_R</th>\n",
       "      <th>PROP_S</th>\n",
       "      <th>RELIABLE_PHENOTYPE</th>\n",
       "      <th>CONSISTENT_PHENOTYPE</th>\n",
       "      <th>IS_NONSYN</th>\n",
       "      <th>IS_MISSENSE</th>\n",
       "      <th>STRUCTURALLY_VALID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>A102R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>S</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>A102T</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>A134V</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>25</td>\n",
       "      <td>0.920000</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>A143G</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>A143T</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>True</td>\n",
       "      <td>S</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>W68S</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>Y103D</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>Y103H</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>Y103S</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>Y34D</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>0.846154</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>195 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "index MUTATION  IS_SNP  IN_CDS   R  S  TOTAL    PROP_R    PROP_S  \\\n",
       "146      A102R    True    True   0  3      3  0.000000  1.000000   \n",
       "147      A102T    True    True   3  0      3  1.000000  0.000000   \n",
       "151      A134V    True    True  23  2     25  0.920000  0.080000   \n",
       "152      A143G    True    True   6  0      6  1.000000  0.000000   \n",
       "154      A143T    True    True   1  6      7  0.142857  0.857143   \n",
       "..         ...     ...     ...  .. ..    ...       ...       ...   \n",
       "536       W68S    True    True   2  0      2  1.000000  0.000000   \n",
       "539      Y103D    True    True   4  0      4  1.000000  0.000000   \n",
       "540      Y103H    True    True   5  0      5  1.000000  0.000000   \n",
       "541      Y103S    True    True   2  0      2  1.000000  0.000000   \n",
       "545       Y34D    True    True  11  2     13  0.846154  0.153846   \n",
       "\n",
       "index  RELIABLE_PHENOTYPE CONSISTENT_PHENOTYPE  IS_NONSYN  IS_MISSENSE  \\\n",
       "146                  True                    S       True         True   \n",
       "147                  True                    R       True         True   \n",
       "151                  True                    R       True         True   \n",
       "152                  True                    R       True         True   \n",
       "154                  True                    S       True         True   \n",
       "..                    ...                  ...        ...          ...   \n",
       "536                  True                    R       True         True   \n",
       "539                  True                    R       True         True   \n",
       "540                  True                    R       True         True   \n",
       "541                  True                    R       True         True   \n",
       "545                  True                    R       True         True   \n",
       "\n",
       "index  STRUCTURALLY_VALID  \n",
       "146                  True  \n",
       "147                  True  \n",
       "151                  True  \n",
       "152                  True  \n",
       "154                  True  \n",
       "..                    ...  \n",
       "536                  True  \n",
       "539                  True  \n",
       "540                  True  \n",
       "541                  True  \n",
       "545                  True  \n",
       "\n",
       "[195 rows x 13 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATASET.rename(columns={'PHENOTYPE':'CONSISTENT_PHENOTYPE'}, inplace=True)\n",
    "DATASET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_88397/692473263.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  DATASET['SEGID']='A'\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "R    155\n",
       "S     40\n",
       "Name: CONSISTENT_PHENOTYPE, dtype: int64"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATASET.loc[DATASET.CONSISTENT_PHENOTYPE.notna()].to_csv(filestem+'-full.csv',index=False)\n",
    "\n",
    "DATASET.loc[DATASET.CONSISTENT_PHENOTYPE.notna()][['MUTATION','CONSISTENT_PHENOTYPE']].to_csv(filestem+'-phen.csv',index=False)\n",
    "\n",
    "DATASET.loc[(DATASET.CONSISTENT_PHENOTYPE.notna())][['MUTATION']].to_csv(filestem+'-muts.csv',index=False, header=False)\n",
    "\n",
    "DATASET['SEGID']='A'\n",
    "DATASET.loc[(DATASET.CONSISTENT_PHENOTYPE.notna())][['SEGID','MUTATION']].to_csv(filestem+'-semu.csv',index=False, header=False, sep=' ')\n",
    "\n",
    "DATASET.CONSISTENT_PHENOTYPE.value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>index</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>TOTAL</th>\n",
       "      <th>PROP_R</th>\n",
       "      <th>PROP_S</th>\n",
       "      <th>RELIABLE_PHENOTYPE</th>\n",
       "      <th>CONSISTENT_PHENOTYPE</th>\n",
       "      <th>IS_NONSYN</th>\n",
       "      <th>IS_MISSENSE</th>\n",
       "      <th>STRUCTURALLY_VALID</th>\n",
       "      <th>SEGID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>A102R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>S</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>A102T</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>A134V</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>25</td>\n",
       "      <td>0.920000</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>A143G</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>A143T</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>True</td>\n",
       "      <td>S</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>W68S</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>Y103D</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>Y103H</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>Y103S</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>Y34D</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>0.846154</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>195 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "index MUTATION  IS_SNP  IN_CDS   R  S  TOTAL    PROP_R    PROP_S  \\\n",
       "146      A102R    True    True   0  3      3  0.000000  1.000000   \n",
       "147      A102T    True    True   3  0      3  1.000000  0.000000   \n",
       "151      A134V    True    True  23  2     25  0.920000  0.080000   \n",
       "152      A143G    True    True   6  0      6  1.000000  0.000000   \n",
       "154      A143T    True    True   1  6      7  0.142857  0.857143   \n",
       "..         ...     ...     ...  .. ..    ...       ...       ...   \n",
       "536       W68S    True    True   2  0      2  1.000000  0.000000   \n",
       "539      Y103D    True    True   4  0      4  1.000000  0.000000   \n",
       "540      Y103H    True    True   5  0      5  1.000000  0.000000   \n",
       "541      Y103S    True    True   2  0      2  1.000000  0.000000   \n",
       "545       Y34D    True    True  11  2     13  0.846154  0.153846   \n",
       "\n",
       "index  RELIABLE_PHENOTYPE CONSISTENT_PHENOTYPE  IS_NONSYN  IS_MISSENSE  \\\n",
       "146                  True                    S       True         True   \n",
       "147                  True                    R       True         True   \n",
       "151                  True                    R       True         True   \n",
       "152                  True                    R       True         True   \n",
       "154                  True                    S       True         True   \n",
       "..                    ...                  ...        ...          ...   \n",
       "536                  True                    R       True         True   \n",
       "539                  True                    R       True         True   \n",
       "540                  True                    R       True         True   \n",
       "541                  True                    R       True         True   \n",
       "545                  True                    R       True         True   \n",
       "\n",
       "index  STRUCTURALLY_VALID SEGID  \n",
       "146                  True     A  \n",
       "147                  True     A  \n",
       "151                  True     A  \n",
       "152                  True     A  \n",
       "154                  True     A  \n",
       "..                    ...   ...  \n",
       "536                  True     A  \n",
       "539                  True     A  \n",
       "540                  True     A  \n",
       "541                  True     A  \n",
       "545                  True     A  \n",
       "\n",
       "[195 rows x 14 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATASET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "DST = pandas.read_csv('data/source-data/cryptic1-dst.csv')\n",
    "MUTATIONS = pandas.read_csv('data/source-data/cryptic1-mutations.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>DRUG</th>\n",
       "      <th>SOURCE</th>\n",
       "      <th>METHOD_1</th>\n",
       "      <th>METHOD_2</th>\n",
       "      <th>METHOD_3</th>\n",
       "      <th>METHOD_CC</th>\n",
       "      <th>METHOD_MIC</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.24.subj.PT-2.lab.1997-606.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>SEQTREAT2020</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>BACTEC460</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.24.subj.PT-4.lab.1998-151.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>SEQTREAT2020</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>BACTEC460</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.24.subj.PT-5.lab.1998-713.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>SEQTREAT2020</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>BACTEC460</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>site.24.subj.PT-7.lab.1999-097.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>SEQTREAT2020</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>BACTEC460</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>site.24.subj.PT-8.lab.1999-131.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>SEQTREAT2020</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>BACTEC460</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22837</th>\n",
       "      <td>site.00.subj.LE10KTB_21.lab.7627886.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>NEJM2018</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22838</th>\n",
       "      <td>site.00.subj.LE10KTB_8.lab.7627900.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>NEJM2018</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22839</th>\n",
       "      <td>site.00.subj.LE10KTB_12.lab.7628121.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>NEJM2018</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22840</th>\n",
       "      <td>site.00.subj.LE10KTB_14.lab.7628143.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>NEJM2018</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22841</th>\n",
       "      <td>site.00.subj.LE10KTB_9.lab.7628147.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>NEJM2018</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22842 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        UNIQUEID DRUG        SOURCE  \\\n",
       "0           site.24.subj.PT-2.lab.1997-606.iso.1  PZA  SEQTREAT2020   \n",
       "1           site.24.subj.PT-4.lab.1998-151.iso.1  PZA  SEQTREAT2020   \n",
       "2           site.24.subj.PT-5.lab.1998-713.iso.1  PZA  SEQTREAT2020   \n",
       "3           site.24.subj.PT-7.lab.1999-097.iso.1  PZA  SEQTREAT2020   \n",
       "4           site.24.subj.PT-8.lab.1999-131.iso.1  PZA  SEQTREAT2020   \n",
       "...                                          ...  ...           ...   \n",
       "22837  site.00.subj.LE10KTB_21.lab.7627886.iso.1  PZA      NEJM2018   \n",
       "22838   site.00.subj.LE10KTB_8.lab.7627900.iso.1  PZA      NEJM2018   \n",
       "22839  site.00.subj.LE10KTB_12.lab.7628121.iso.1  PZA      NEJM2018   \n",
       "22840  site.00.subj.LE10KTB_14.lab.7628143.iso.1  PZA      NEJM2018   \n",
       "22841   site.00.subj.LE10KTB_9.lab.7628147.iso.1  PZA      NEJM2018   \n",
       "\n",
       "           METHOD_1 METHOD_2   METHOD_3 METHOD_CC  METHOD_MIC PHENOTYPE  \n",
       "0      liquid media     MGIT  BACTEC460     100.0         NaN         S  \n",
       "1      liquid media     MGIT  BACTEC460     100.0         NaN         S  \n",
       "2      liquid media     MGIT  BACTEC460     100.0         NaN         S  \n",
       "3      liquid media     MGIT  BACTEC460     100.0         NaN         S  \n",
       "4      liquid media     MGIT  BACTEC460     100.0         NaN         S  \n",
       "...             ...      ...        ...       ...         ...       ...  \n",
       "22837  liquid media     MGIT        NaN       NaN         NaN         S  \n",
       "22838  liquid media     MGIT        NaN       NaN         NaN         S  \n",
       "22839  liquid media     MGIT        NaN       NaN         NaN         S  \n",
       "22840  liquid media     MGIT        NaN       NaN         NaN         S  \n",
       "22841  liquid media     MGIT        NaN       NaN         NaN         S  \n",
       "\n",
       "[22842 rows x 9 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3549"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(MUTATIONS.UNIQUEID.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3573"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(MUTATIONS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3331"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(MUTATIONS[MUTATIONS.AMINO_ACID_NUMBER.notna()].UNIQUEID.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "foo = MUTATIONS[MUTATIONS.AMINO_ACID_NUMBER.notna()][['UNIQUEID', 'POSITION']].groupby('UNIQUEID').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    3314\n",
       "2      17\n",
       "Name: POSITION, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "foo.POSITION.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "singles = foo[foo.POSITION==1].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    2911\n",
       "True      403\n",
       "Name: IS_INDEL, dtype: int64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "MUTATIONS[(MUTATIONS.AMINO_ACID_NUMBER.notna()) & (MUTATIONS.UNIQUEID.isin(singles))].IS_INDEL.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "indels = MUTATIONS[(MUTATIONS.AMINO_ACID_NUMBER.notna()) & (MUTATIONS.UNIQUEID.isin(singles)) & (MUTATIONS.IS_INDEL | MUTATIONS.IN_PROMOTER)].UNIQUEID.unique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([], dtype=object)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS[(MUTATIONS.AMINO_ACID_NUMBER.notna()) & (MUTATIONS.UNIQUEID.isin(singles)) & (MUTATIONS.IN_PROMOTER)].UNIQUEID.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>IN_PROMOTER</th>\n",
       "      <th>IS_SYNONYMOUS</th>\n",
       "      <th>IS_NONSYNONYMOUS</th>\n",
       "      <th>IS_HET</th>\n",
       "      <th>IS_NULL</th>\n",
       "      <th>IS_FILTER_PASS</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.05.subj.PSLM-0791.lab.SLM-049.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>L120P</td>\n",
       "      <td>120.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctg</td>\n",
       "      <td>ccg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.05.subj.LR-2032.lab.FN-00407-15.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>W119L</td>\n",
       "      <td>119.0</td>\n",
       "      <td>119.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tgg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.05.subj.PMK-1015.lab.MK-1781.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>F58L</td>\n",
       "      <td>58.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ttc</td>\n",
       "      <td>ctc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>site.05.subj.LR-2417.lab.FN-01304-17.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>H51R</td>\n",
       "      <td>51.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cac</td>\n",
       "      <td>cgc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>site.05.subj.LR-2162.lab.FN-00284-16.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>V139L</td>\n",
       "      <td>139.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gtg</td>\n",
       "      <td>ctg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3568</th>\n",
       "      <td>site.10.subj.YA00125758.lab.YA00125758.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>467_indel</td>\n",
       "      <td>467.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>2288775.0</td>\n",
       "      <td>467.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>1.0</td>\n",
       "      <td>467_ins</td>\n",
       "      <td>467_ins_1</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3569</th>\n",
       "      <td>site.10.subj.SADG00497215_S8.lab.DG00497215_S8...</td>\n",
       "      <td>pncA</td>\n",
       "      <td>C138R</td>\n",
       "      <td>138.0</td>\n",
       "      <td>138.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tgt</td>\n",
       "      <td>cgt</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3570</th>\n",
       "      <td>site.10.subj.KD01666167.lab.KD01666167.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>H71Y</td>\n",
       "      <td>71.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cat</td>\n",
       "      <td>tat</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3571</th>\n",
       "      <td>site.10.subj.YA00029870.lab.YA00029870.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>a-11g</td>\n",
       "      <td>-11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2289252.0</td>\n",
       "      <td>-11.0</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3572</th>\n",
       "      <td>site.10.subj.YA00008913.lab.YA00008913.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>G97D</td>\n",
       "      <td>97.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ggt</td>\n",
       "      <td>gat</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3573 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               UNIQUEID  GENE   MUTATION  \\\n",
       "0              site.05.subj.PSLM-0791.lab.SLM-049.iso.1  pncA      L120P   \n",
       "1            site.05.subj.LR-2032.lab.FN-00407-15.iso.1  pncA      W119L   \n",
       "2               site.05.subj.PMK-1015.lab.MK-1781.iso.1  pncA       F58L   \n",
       "3            site.05.subj.LR-2417.lab.FN-01304-17.iso.1  pncA       H51R   \n",
       "4            site.05.subj.LR-2162.lab.FN-00284-16.iso.1  pncA      V139L   \n",
       "...                                                 ...   ...        ...   \n",
       "3568       site.10.subj.YA00125758.lab.YA00125758.iso.1  pncA  467_indel   \n",
       "3569  site.10.subj.SADG00497215_S8.lab.DG00497215_S8...  pncA      C138R   \n",
       "3570       site.10.subj.KD01666167.lab.KD01666167.iso.1  pncA       H71Y   \n",
       "3571       site.10.subj.YA00029870.lab.YA00029870.iso.1  pncA      a-11g   \n",
       "3572       site.10.subj.YA00008913.lab.YA00008913.iso.1  pncA       G97D   \n",
       "\n",
       "      POSITION  AMINO_ACID_NUMBER  GENOME_INDEX  NUCLEOTIDE_NUMBER  REF  ALT  \\\n",
       "0        120.0              120.0           NaN                NaN  ctg  ccg   \n",
       "1        119.0              119.0           NaN                NaN  tgg  ttg   \n",
       "2         58.0               58.0           NaN                NaN  ttc  ctc   \n",
       "3         51.0               51.0           NaN                NaN  cac  cgc   \n",
       "4        139.0              139.0           NaN                NaN  gtg  ctg   \n",
       "...        ...                ...           ...                ...  ...  ...   \n",
       "3568     467.0              156.0     2288775.0              467.0  NaN  NaN   \n",
       "3569     138.0              138.0           NaN                NaN  tgt  cgt   \n",
       "3570      71.0               71.0           NaN                NaN  cat  tat   \n",
       "3571     -11.0                NaN     2289252.0              -11.0    a    g   \n",
       "3572      97.0               97.0           NaN                NaN  ggt  gat   \n",
       "\n",
       "      IS_SNP  IS_INDEL  IN_CDS  IN_PROMOTER  IS_SYNONYMOUS  IS_NONSYNONYMOUS  \\\n",
       "0       True     False    True        False          False              True   \n",
       "1       True     False    True        False          False              True   \n",
       "2       True     False    True        False          False              True   \n",
       "3       True     False    True        False          False              True   \n",
       "4       True     False    True        False          False              True   \n",
       "...      ...       ...     ...          ...            ...               ...   \n",
       "3568   False      True    True        False          False             False   \n",
       "3569    True     False    True        False          False              True   \n",
       "3570    True     False    True        False          False              True   \n",
       "3571    True     False   False         True          False             False   \n",
       "3572    True     False    True        False          False              True   \n",
       "\n",
       "      IS_HET  IS_NULL  IS_FILTER_PASS ELEMENT_TYPE MUTATION_TYPE  \\\n",
       "0      False    False            True         GENE           AAM   \n",
       "1      False    False            True         GENE           AAM   \n",
       "2      False    False            True         GENE           AAM   \n",
       "3      False    False            True         GENE           AAM   \n",
       "4      False    False            True         GENE           AAM   \n",
       "...      ...      ...             ...          ...           ...   \n",
       "3568   False    False            True         GENE         INDEL   \n",
       "3569   False    False            True         GENE           AAM   \n",
       "3570   False    False            True         GENE           AAM   \n",
       "3571   False    False            True         GENE           SNP   \n",
       "3572   False    False            True         GENE           AAM   \n",
       "\n",
       "      INDEL_LENGTH  INDEL_1    INDEL_2 SITEID  NUMBER_NUCLEOTIDE_CHANGES  \n",
       "0              NaN      NaN        NaN     05                          1  \n",
       "1              NaN      NaN        NaN     05                          1  \n",
       "2              NaN      NaN        NaN     05                          1  \n",
       "3              NaN      NaN        NaN     05                          1  \n",
       "4              NaN      NaN        NaN     05                          1  \n",
       "...            ...      ...        ...    ...                        ...  \n",
       "3568           1.0  467_ins  467_ins_1     10                          0  \n",
       "3569           NaN      NaN        NaN     10                          1  \n",
       "3570           NaN      NaN        NaN     10                          1  \n",
       "3571           NaN      NaN        NaN     10                          0  \n",
       "3572           NaN      NaN        NaN     10                          1  \n",
       "\n",
       "[3573 rows x 25 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R    355\n",
       "S     48\n",
       "Name: PHENOTYPE, dtype: int64"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "DST[DST.UNIQUEID.isin(indels)].PHENOTYPE.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "7a792fcb311f9eb9f3c1b942a8c87ada8484712b89b670347c16a1088e0a1f69"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
