{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sbmlcore, pandas, numpy, copy\n",
    "import itertools\n",
    "\n",
    "pandas.options.display.max_columns=999"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in three sets of clinical samples with mutations and recorded phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "miotto2014 199 755\n",
      "whitfield2015 65 634\n",
      "cryptic2021 526 3525\n",
      "('miotto2014', 'whitfield2015') 41\n",
      "('miotto2014', 'cryptic2021') 178\n",
      "('whitfield2015', 'cryptic2021') 46\n",
      "('miotto2014', 'whitfield2015', 'cryptic2021') 35\n"
     ]
    }
   ],
   "source": [
    "filestem = 'data/ds-validation'\n",
    "\n",
    "clinical_sample_csvs = ['miotto2014','whitfield2015','cryptic2021']\n",
    "stem='data/clinical-samples/ds-'\n",
    "clinical_sample_dfs = {}\n",
    "clinical_sample_sets = {}\n",
    "for i in clinical_sample_csvs:\n",
    "    clinical_sample_dfs[i] = pandas.read_csv(stem+i+'.csv')\n",
    "    clinical_sample_dfs[i].set_index('MUTATION', inplace=True)\n",
    "    clinical_sample_sets[i] = set(clinical_sample_dfs[i].index)\n",
    "    clinical_sample_dfs[i].reset_index(inplace=True)\n",
    "    clinical_sample_dfs[i].set_index(['MUTATION', 'IS_SNP', 'IN_CDS'], inplace=True)\n",
    "    clinical_sample_dfs[i]['total'] = clinical_sample_dfs[i].R + clinical_sample_dfs[i].S\n",
    "\n",
    "for i in clinical_sample_csvs:\n",
    "    print(i, len(clinical_sample_sets[i]), clinical_sample_dfs[i].total.sum())\n",
    "if len(clinical_sample_csvs)>1:\n",
    "    for i in list(itertools.combinations(clinical_sample_csvs,2)):\n",
    "        print(i, len(clinical_sample_sets[i[0]].intersection(clinical_sample_sets[i[1]])))\n",
    "if len(clinical_sample_csvs)>2:\n",
    "    for i in list(itertools.combinations(clinical_sample_csvs,3)):\n",
    "        print(i, len(clinical_sample_sets[i[0]] & clinical_sample_sets[i[1]] & clinical_sample_sets[i[2]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>index</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>TOTAL</th>\n",
       "      <th>PROP_R</th>\n",
       "      <th>PROP_S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!187G</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!187R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-29_indel</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "index   MUTATION  IS_SNP  IN_CDS  R  S  TOTAL  PROP_R  PROP_S\n",
       "0          !187G    True    True  1  1      2     0.5     0.5\n",
       "1          !187R    True    True  0  1      1     0.0     1.0\n",
       "2      -29_indel   False   False  0  1      1     0.0     1.0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINCAL_SAMPLES = clinical_sample_dfs['cryptic2021'].join(clinical_sample_dfs['miotto2014'], lsuffix='s', rsuffix = 'm', how='outer')\n",
    "CLINCAL_SAMPLES = CLINCAL_SAMPLES.join(clinical_sample_dfs['whitfield2015'], rsuffix='w',how='outer')\n",
    "CLINCAL_SAMPLES.rename(columns={'R': 'Rw', 'S': 'Sw', 'total': 'totalw'}, inplace=True)\n",
    "CLINCAL_SAMPLES.fillna(0, inplace=True)   \n",
    "CLINCAL_SAMPLES = CLINCAL_SAMPLES.astype('int')\n",
    "CLINCAL_SAMPLES['R'] = CLINCAL_SAMPLES['Rs'] + CLINCAL_SAMPLES['Rm'] + CLINCAL_SAMPLES['Rw']\n",
    "CLINCAL_SAMPLES['S'] = CLINCAL_SAMPLES['Ss'] + CLINCAL_SAMPLES['Sm'] + CLINCAL_SAMPLES['Sw']\n",
    "CLINCAL_SAMPLES['TOTAL'] = CLINCAL_SAMPLES['R'] + CLINCAL_SAMPLES['S']\n",
    "CLINCAL_SAMPLES.drop(columns=['Rs', 'Ss', 'Rm', 'Sm', 'Rw', 'Sw', 'totals', 'totalm', 'totalw'], inplace=True)\n",
    "CLINCAL_SAMPLES.reset_index(inplace=True)\n",
    "CLINCAL_SAMPLES['PROP_R']=CLINCAL_SAMPLES['R']/CLINCAL_SAMPLES['TOTAL']\n",
    "CLINCAL_SAMPLES['PROP_S']=CLINCAL_SAMPLES['S']/CLINCAL_SAMPLES['TOTAL']\n",
    "CLINCAL_SAMPLES.columns.name='index'\n",
    "CLINCAL_SAMPLES[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>index</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>TOTAL</th>\n",
       "      <th>PROP_R</th>\n",
       "      <th>PROP_S</th>\n",
       "      <th>RELIABLE_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>!187G</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!187R</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-29_indel</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "index      IS_SNP  IN_CDS  R  S  TOTAL  PROP_R  PROP_S  RELIABLE_PHENOTYPE  \\\n",
       "MUTATION                                                                     \n",
       "!187G        True    True  1  1      2     0.5     0.5               False   \n",
       "!187R        True    True  0  1      1     0.0     1.0               False   \n",
       "-29_indel   False   False  0  1      1     0.0     1.0               False   \n",
       "\n",
       "index     PHENOTYPE  \n",
       "MUTATION             \n",
       "!187G             U  \n",
       "!187R             U  \n",
       "-29_indel         U  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def classify(row):\n",
    "    reliable_phenotype=False\n",
    "    phenotype='U'\n",
    "    if row['TOTAL']>=4:\n",
    "        if row['PROP_R']>=0.75:\n",
    "            reliable_phenotype=True\n",
    "            phenotype='R'\n",
    "        elif row['PROP_S']>=0.75:\n",
    "            reliable_phenotype=True\n",
    "            phenotype='S'\n",
    "    elif row['TOTAL']>=2:\n",
    "        if row['R']==row['TOTAL']:\n",
    "            reliable_phenotype=True\n",
    "            phenotype='R'\n",
    "        elif row['S']==row['TOTAL']:\n",
    "            reliable_phenotype=True\n",
    "            phenotype='S'\n",
    "\n",
    "    return pandas.Series([reliable_phenotype, phenotype]) \n",
    "\n",
    "def valid_for_structure(row):\n",
    "    if \"!\" in row.MUTATION:\n",
    "        return False\n",
    "    elif '186' in row.MUTATION:\n",
    "        return False\n",
    "    elif row.MUTATION[0]==row.MUTATION[-1]:\n",
    "        return False\n",
    "    elif not row.IS_SNP:\n",
    "        return False\n",
    "    elif not row.IN_CDS:\n",
    "        return False\n",
    "    else:\n",
    "        return True\n",
    "\n",
    "CLINCAL_SAMPLES[['RELIABLE_PHENOTYPE', 'PHENOTYPE']]=CLINCAL_SAMPLES.apply(classify,axis=1)   \n",
    "CLINCAL_SAMPLES.set_index('MUTATION', inplace=True)\n",
    "CLINCAL_SAMPLES[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R    168\n",
       "S     45\n",
       "Name: PHENOTYPE, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINCAL_SAMPLES[(CLINCAL_SAMPLES.RELIABLE_PHENOTYPE) & (CLINCAL_SAMPLES.IS_SNP) & (CLINCAL_SAMPLES.IN_CDS)].PHENOTYPE.value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_49006/819241938.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  DATASET.rename(columns={'PHENOTYPE':'CONSISTENT_PHENOTYPE'}, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>index</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>TOTAL</th>\n",
       "      <th>PROP_R</th>\n",
       "      <th>PROP_S</th>\n",
       "      <th>RELIABLE_PHENOTYPE</th>\n",
       "      <th>CONSISTENT_PHENOTYPE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A102R</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A102T</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A134V</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>25</td>\n",
       "      <td>0.920000</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A143G</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A143T</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>True</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y34!</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y34D</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>0.846154</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y64!</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y95!</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y99!</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>213 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "index     IS_SNP  IN_CDS   R  S  TOTAL    PROP_R    PROP_S  \\\n",
       "MUTATION                                                     \n",
       "A102R       True    True   0  3      3  0.000000  1.000000   \n",
       "A102T       True    True   3  0      3  1.000000  0.000000   \n",
       "A134V       True    True  23  2     25  0.920000  0.080000   \n",
       "A143G       True    True   6  0      6  1.000000  0.000000   \n",
       "A143T       True    True   1  6      7  0.142857  0.857143   \n",
       "...          ...     ...  .. ..    ...       ...       ...   \n",
       "Y34!        True    True   7  0      7  1.000000  0.000000   \n",
       "Y34D        True    True  11  2     13  0.846154  0.153846   \n",
       "Y64!        True    True  11  0     11  1.000000  0.000000   \n",
       "Y95!        True    True   5  1      6  0.833333  0.166667   \n",
       "Y99!        True    True   4  0      4  1.000000  0.000000   \n",
       "\n",
       "index     RELIABLE_PHENOTYPE CONSISTENT_PHENOTYPE  \n",
       "MUTATION                                           \n",
       "A102R                   True                    S  \n",
       "A102T                   True                    R  \n",
       "A134V                   True                    R  \n",
       "A143G                   True                    R  \n",
       "A143T                   True                    S  \n",
       "...                      ...                  ...  \n",
       "Y34!                    True                    R  \n",
       "Y34D                    True                    R  \n",
       "Y64!                    True                    R  \n",
       "Y95!                    True                    R  \n",
       "Y99!                    True                    R  \n",
       "\n",
       "[213 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATASET = CLINCAL_SAMPLES[(CLINCAL_SAMPLES.RELIABLE_PHENOTYPE) & (CLINCAL_SAMPLES.IS_SNP) & (CLINCAL_SAMPLES.IN_CDS)]\n",
    "DATASET.rename(columns={'PHENOTYPE':'CONSISTENT_PHENOTYPE'}, inplace=True)\n",
    "DATASET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_49006/3242486175.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  DATASET['IS_SYN'] = DATASET.apply(syn, axis=1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>index</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>TOTAL</th>\n",
       "      <th>PROP_R</th>\n",
       "      <th>PROP_S</th>\n",
       "      <th>RELIABLE_PHENOTYPE</th>\n",
       "      <th>CONSISTENT_PHENOTYPE</th>\n",
       "      <th>IS_SYN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A102R</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>True</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A102T</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A134V</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>25</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.08</td>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "index     IS_SNP  IN_CDS   R  S  TOTAL  PROP_R  PROP_S  RELIABLE_PHENOTYPE  \\\n",
       "MUTATION                                                                     \n",
       "A102R       True    True   0  3      3    0.00    1.00                True   \n",
       "A102T       True    True   3  0      3    1.00    0.00                True   \n",
       "A134V       True    True  23  2     25    0.92    0.08                True   \n",
       "\n",
       "index    CONSISTENT_PHENOTYPE  IS_SYN  \n",
       "MUTATION                               \n",
       "A102R                       S   False  \n",
       "A102T                       R   False  \n",
       "A134V                       R   False  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def syn(row):\n",
    "    if row.MUTATION[0]==row.MUTATION[-1]:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "DATASET.reset_index(inplace=True)\n",
    "DATASET['IS_SYN'] = DATASET.apply(syn, axis=1)\n",
    "DATASET.set_index('MUTATION',inplace=True)\n",
    "DATASET = DATASET.loc[~DATASET.IS_SYN]\n",
    "DATASET[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R    168\n",
       "S     44\n",
       "Name: CONSISTENT_PHENOTYPE, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATASET.CONSISTENT_PHENOTYPE.value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_49006/92460833.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  DATASET['STRUCTURALLY_VALID'] = DATASET.apply(valid_for_structure, axis=1)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "R    155\n",
       "S     44\n",
       "Name: CONSISTENT_PHENOTYPE, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATASET.reset_index(inplace=True)\n",
    "\n",
    "def valid_for_structure(row):\n",
    "    if \"!\" in row.MUTATION:\n",
    "        return False\n",
    "    elif '186' in row.MUTATION:\n",
    "        return False\n",
    "    elif row.MUTATION[0]==row.MUTATION[-1]:\n",
    "        return False\n",
    "    elif not row.IS_SNP:\n",
    "        return False\n",
    "    elif not row.IN_CDS:\n",
    "        return False\n",
    "    else:\n",
    "        return True\n",
    "\n",
    "DATASET['STRUCTURALLY_VALID'] = DATASET.apply(valid_for_structure, axis=1)\n",
    "DATASET=DATASET[DATASET.STRUCTURALLY_VALID]\n",
    "DATASET.drop(columns=['STRUCTURALLY_VALID'],inplace=True)\n",
    "\n",
    "DATASET.loc[DATASET.CONSISTENT_PHENOTYPE.notna()].to_csv(filestem+'-full.csv',index=False)\n",
    "\n",
    "DATASET.loc[DATASET.CONSISTENT_PHENOTYPE.notna()][['MUTATION','CONSISTENT_PHENOTYPE']].to_csv(filestem+'-phen.csv',index=False)\n",
    "\n",
    "DATASET.loc[(DATASET.CONSISTENT_PHENOTYPE.notna())][['MUTATION']].to_csv(filestem+'-muts.csv',index=False, header=False)\n",
    "\n",
    "DATASET['SEGID']='A'\n",
    "DATASET.loc[(DATASET.CONSISTENT_PHENOTYPE.notna())][['SEGID','MUTATION']].to_csv(filestem+'-semu.csv',index=False, header=False, sep=' ')\n",
    "\n",
    "DATASET.CONSISTENT_PHENOTYPE.value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "DST = pandas.read_csv('data/source-data/cryptic1-dst.csv')\n",
    "MUTATIONS = pandas.read_csv('data/source-data/cryptic1-mutations.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>DRUG</th>\n",
       "      <th>SOURCE</th>\n",
       "      <th>METHOD_1</th>\n",
       "      <th>METHOD_2</th>\n",
       "      <th>METHOD_3</th>\n",
       "      <th>METHOD_CC</th>\n",
       "      <th>METHOD_MIC</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.24.subj.PT-2.lab.1997-606.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>SEQTREAT2020</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>BACTEC460</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.24.subj.PT-4.lab.1998-151.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>SEQTREAT2020</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>BACTEC460</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.24.subj.PT-5.lab.1998-713.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>SEQTREAT2020</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>BACTEC460</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>site.24.subj.PT-7.lab.1999-097.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>SEQTREAT2020</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>BACTEC460</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>site.24.subj.PT-8.lab.1999-131.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>SEQTREAT2020</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>BACTEC460</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22837</th>\n",
       "      <td>site.00.subj.LE10KTB_21.lab.7627886.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>NEJM2018</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22838</th>\n",
       "      <td>site.00.subj.LE10KTB_8.lab.7627900.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>NEJM2018</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22839</th>\n",
       "      <td>site.00.subj.LE10KTB_12.lab.7628121.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>NEJM2018</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22840</th>\n",
       "      <td>site.00.subj.LE10KTB_14.lab.7628143.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>NEJM2018</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22841</th>\n",
       "      <td>site.00.subj.LE10KTB_9.lab.7628147.iso.1</td>\n",
       "      <td>PZA</td>\n",
       "      <td>NEJM2018</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>MGIT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22842 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        UNIQUEID DRUG        SOURCE  \\\n",
       "0           site.24.subj.PT-2.lab.1997-606.iso.1  PZA  SEQTREAT2020   \n",
       "1           site.24.subj.PT-4.lab.1998-151.iso.1  PZA  SEQTREAT2020   \n",
       "2           site.24.subj.PT-5.lab.1998-713.iso.1  PZA  SEQTREAT2020   \n",
       "3           site.24.subj.PT-7.lab.1999-097.iso.1  PZA  SEQTREAT2020   \n",
       "4           site.24.subj.PT-8.lab.1999-131.iso.1  PZA  SEQTREAT2020   \n",
       "...                                          ...  ...           ...   \n",
       "22837  site.00.subj.LE10KTB_21.lab.7627886.iso.1  PZA      NEJM2018   \n",
       "22838   site.00.subj.LE10KTB_8.lab.7627900.iso.1  PZA      NEJM2018   \n",
       "22839  site.00.subj.LE10KTB_12.lab.7628121.iso.1  PZA      NEJM2018   \n",
       "22840  site.00.subj.LE10KTB_14.lab.7628143.iso.1  PZA      NEJM2018   \n",
       "22841   site.00.subj.LE10KTB_9.lab.7628147.iso.1  PZA      NEJM2018   \n",
       "\n",
       "           METHOD_1 METHOD_2   METHOD_3 METHOD_CC  METHOD_MIC PHENOTYPE  \n",
       "0      liquid media     MGIT  BACTEC460     100.0         NaN         S  \n",
       "1      liquid media     MGIT  BACTEC460     100.0         NaN         S  \n",
       "2      liquid media     MGIT  BACTEC460     100.0         NaN         S  \n",
       "3      liquid media     MGIT  BACTEC460     100.0         NaN         S  \n",
       "4      liquid media     MGIT  BACTEC460     100.0         NaN         S  \n",
       "...             ...      ...        ...       ...         ...       ...  \n",
       "22837  liquid media     MGIT        NaN       NaN         NaN         S  \n",
       "22838  liquid media     MGIT        NaN       NaN         NaN         S  \n",
       "22839  liquid media     MGIT        NaN       NaN         NaN         S  \n",
       "22840  liquid media     MGIT        NaN       NaN         NaN         S  \n",
       "22841  liquid media     MGIT        NaN       NaN         NaN         S  \n",
       "\n",
       "[22842 rows x 9 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3549"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(MUTATIONS.UNIQUEID.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3573"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(MUTATIONS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3331"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(MUTATIONS[MUTATIONS.AMINO_ACID_NUMBER.notna()].UNIQUEID.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "foo = MUTATIONS[MUTATIONS.AMINO_ACID_NUMBER.notna()][['UNIQUEID', 'POSITION']].groupby('UNIQUEID').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    3314\n",
       "2      17\n",
       "Name: POSITION, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "foo.POSITION.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "singles = foo[foo.POSITION==1].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    2911\n",
       "True      403\n",
       "Name: IS_INDEL, dtype: int64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "MUTATIONS[(MUTATIONS.AMINO_ACID_NUMBER.notna()) & (MUTATIONS.UNIQUEID.isin(singles))].IS_INDEL.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "indels = MUTATIONS[(MUTATIONS.AMINO_ACID_NUMBER.notna()) & (MUTATIONS.UNIQUEID.isin(singles)) & (MUTATIONS.IS_INDEL | MUTATIONS.IN_PROMOTER)].UNIQUEID.unique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([], dtype=object)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS[(MUTATIONS.AMINO_ACID_NUMBER.notna()) & (MUTATIONS.UNIQUEID.isin(singles)) & (MUTATIONS.IN_PROMOTER)].UNIQUEID.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>IN_PROMOTER</th>\n",
       "      <th>IS_SYNONYMOUS</th>\n",
       "      <th>IS_NONSYNONYMOUS</th>\n",
       "      <th>IS_HET</th>\n",
       "      <th>IS_NULL</th>\n",
       "      <th>IS_FILTER_PASS</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.05.subj.PSLM-0791.lab.SLM-049.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>L120P</td>\n",
       "      <td>120.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctg</td>\n",
       "      <td>ccg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.05.subj.LR-2032.lab.FN-00407-15.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>W119L</td>\n",
       "      <td>119.0</td>\n",
       "      <td>119.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tgg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.05.subj.PMK-1015.lab.MK-1781.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>F58L</td>\n",
       "      <td>58.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ttc</td>\n",
       "      <td>ctc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>site.05.subj.LR-2417.lab.FN-01304-17.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>H51R</td>\n",
       "      <td>51.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cac</td>\n",
       "      <td>cgc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>site.05.subj.LR-2162.lab.FN-00284-16.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>V139L</td>\n",
       "      <td>139.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gtg</td>\n",
       "      <td>ctg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3568</th>\n",
       "      <td>site.10.subj.YA00125758.lab.YA00125758.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>467_indel</td>\n",
       "      <td>467.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>2288775.0</td>\n",
       "      <td>467.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>1.0</td>\n",
       "      <td>467_ins</td>\n",
       "      <td>467_ins_1</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3569</th>\n",
       "      <td>site.10.subj.SADG00497215_S8.lab.DG00497215_S8...</td>\n",
       "      <td>pncA</td>\n",
       "      <td>C138R</td>\n",
       "      <td>138.0</td>\n",
       "      <td>138.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tgt</td>\n",
       "      <td>cgt</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3570</th>\n",
       "      <td>site.10.subj.KD01666167.lab.KD01666167.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>H71Y</td>\n",
       "      <td>71.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cat</td>\n",
       "      <td>tat</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3571</th>\n",
       "      <td>site.10.subj.YA00029870.lab.YA00029870.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>a-11g</td>\n",
       "      <td>-11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2289252.0</td>\n",
       "      <td>-11.0</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>SNP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3572</th>\n",
       "      <td>site.10.subj.YA00008913.lab.YA00008913.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>G97D</td>\n",
       "      <td>97.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ggt</td>\n",
       "      <td>gat</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3573 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               UNIQUEID  GENE   MUTATION  \\\n",
       "0              site.05.subj.PSLM-0791.lab.SLM-049.iso.1  pncA      L120P   \n",
       "1            site.05.subj.LR-2032.lab.FN-00407-15.iso.1  pncA      W119L   \n",
       "2               site.05.subj.PMK-1015.lab.MK-1781.iso.1  pncA       F58L   \n",
       "3            site.05.subj.LR-2417.lab.FN-01304-17.iso.1  pncA       H51R   \n",
       "4            site.05.subj.LR-2162.lab.FN-00284-16.iso.1  pncA      V139L   \n",
       "...                                                 ...   ...        ...   \n",
       "3568       site.10.subj.YA00125758.lab.YA00125758.iso.1  pncA  467_indel   \n",
       "3569  site.10.subj.SADG00497215_S8.lab.DG00497215_S8...  pncA      C138R   \n",
       "3570       site.10.subj.KD01666167.lab.KD01666167.iso.1  pncA       H71Y   \n",
       "3571       site.10.subj.YA00029870.lab.YA00029870.iso.1  pncA      a-11g   \n",
       "3572       site.10.subj.YA00008913.lab.YA00008913.iso.1  pncA       G97D   \n",
       "\n",
       "      POSITION  AMINO_ACID_NUMBER  GENOME_INDEX  NUCLEOTIDE_NUMBER  REF  ALT  \\\n",
       "0        120.0              120.0           NaN                NaN  ctg  ccg   \n",
       "1        119.0              119.0           NaN                NaN  tgg  ttg   \n",
       "2         58.0               58.0           NaN                NaN  ttc  ctc   \n",
       "3         51.0               51.0           NaN                NaN  cac  cgc   \n",
       "4        139.0              139.0           NaN                NaN  gtg  ctg   \n",
       "...        ...                ...           ...                ...  ...  ...   \n",
       "3568     467.0              156.0     2288775.0              467.0  NaN  NaN   \n",
       "3569     138.0              138.0           NaN                NaN  tgt  cgt   \n",
       "3570      71.0               71.0           NaN                NaN  cat  tat   \n",
       "3571     -11.0                NaN     2289252.0              -11.0    a    g   \n",
       "3572      97.0               97.0           NaN                NaN  ggt  gat   \n",
       "\n",
       "      IS_SNP  IS_INDEL  IN_CDS  IN_PROMOTER  IS_SYNONYMOUS  IS_NONSYNONYMOUS  \\\n",
       "0       True     False    True        False          False              True   \n",
       "1       True     False    True        False          False              True   \n",
       "2       True     False    True        False          False              True   \n",
       "3       True     False    True        False          False              True   \n",
       "4       True     False    True        False          False              True   \n",
       "...      ...       ...     ...          ...            ...               ...   \n",
       "3568   False      True    True        False          False             False   \n",
       "3569    True     False    True        False          False              True   \n",
       "3570    True     False    True        False          False              True   \n",
       "3571    True     False   False         True          False             False   \n",
       "3572    True     False    True        False          False              True   \n",
       "\n",
       "      IS_HET  IS_NULL  IS_FILTER_PASS ELEMENT_TYPE MUTATION_TYPE  \\\n",
       "0      False    False            True         GENE           AAM   \n",
       "1      False    False            True         GENE           AAM   \n",
       "2      False    False            True         GENE           AAM   \n",
       "3      False    False            True         GENE           AAM   \n",
       "4      False    False            True         GENE           AAM   \n",
       "...      ...      ...             ...          ...           ...   \n",
       "3568   False    False            True         GENE         INDEL   \n",
       "3569   False    False            True         GENE           AAM   \n",
       "3570   False    False            True         GENE           AAM   \n",
       "3571   False    False            True         GENE           SNP   \n",
       "3572   False    False            True         GENE           AAM   \n",
       "\n",
       "      INDEL_LENGTH  INDEL_1    INDEL_2 SITEID  NUMBER_NUCLEOTIDE_CHANGES  \n",
       "0              NaN      NaN        NaN     05                          1  \n",
       "1              NaN      NaN        NaN     05                          1  \n",
       "2              NaN      NaN        NaN     05                          1  \n",
       "3              NaN      NaN        NaN     05                          1  \n",
       "4              NaN      NaN        NaN     05                          1  \n",
       "...            ...      ...        ...    ...                        ...  \n",
       "3568           1.0  467_ins  467_ins_1     10                          0  \n",
       "3569           NaN      NaN        NaN     10                          1  \n",
       "3570           NaN      NaN        NaN     10                          1  \n",
       "3571           NaN      NaN        NaN     10                          0  \n",
       "3572           NaN      NaN        NaN     10                          1  \n",
       "\n",
       "[3573 rows x 25 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R    355\n",
       "S     48\n",
       "Name: PHENOTYPE, dtype: int64"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "DST[DST.UNIQUEID.isin(indels)].PHENOTYPE.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "7a792fcb311f9eb9f3c1b942a8c87ada8484712b89b670347c16a1088e0a1f69"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
