{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy, pandas, pickle\n",
    "\n",
    "import gumpy, joblib, numpy\n",
    "\n",
    "from skops.io import load\n",
    "\n",
    "from misc import construct_line, DataFrameSelector\n",
    "\n",
    "pandas.options.display.max_columns = 999"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the TB reference and build a pncA gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref = gumpy.Genome('/Users/fowler/packages/tuberculosis_amr_catalogues/catalogues/NC_000962.3/NC_000962.3.gbk')\n",
    "pnca = ref.build_gene('pncA')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Iterate through the nucleotide sequence of the gene to find all the single SNP missense mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "bases = ['a', 't', 'c', 'g']\n",
    "\n",
    "mutations = []\n",
    "\n",
    "for idx,nuc in zip(pnca.nucleotide_index, pnca.nucleotide_sequence):\n",
    "\n",
    "    for j in bases:\n",
    "\n",
    "        if nuc==j:\n",
    "            continue\n",
    "\n",
    "        # make the single SNP to a copy\n",
    "        new_pnca = copy.deepcopy(pnca)\n",
    "        new_pnca.nucleotide_sequence[new_pnca.nucleotide_index==idx]=j\n",
    "\n",
    "        # translate \n",
    "        new_pnca._translate_sequence()\n",
    "\n",
    "        # find out the resulting mutation\n",
    "        gene_diff = pnca - new_pnca\n",
    "        mut = gene_diff.mutations[0]\n",
    "\n",
    "        # ignore promoter mutations\n",
    "        if '-' in mut:\n",
    "            continue\n",
    "\n",
    "        # ignore synoymouse mutations\n",
    "        if mut[0]==mut[-1]:\n",
    "            continue\n",
    "        \n",
    "        # ignore nonsense mutations\n",
    "        if mut[0]=='!':\n",
    "            continue\n",
    "        if mut[-1]=='!':\n",
    "            continue\n",
    "\n",
    "        # skip these as they are not resolved in the protein structure\n",
    "        if mut[1:-1]=='186':\n",
    "            continue\n",
    "\n",
    "        # if we haven't seen this mutation before, remember it\n",
    "        if mut not in mutations:\n",
    "            mutations.append(mut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 1099 missense mutations.\n"
     ]
    }
   ],
   "source": [
    "print(\"There are %i missense mutations.\" % len(mutations))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>CONSISTENT_PHENOTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1L</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1V</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M1K</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M1T</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M1R</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  MUTATION CONSISTENT_PHENOTYPE\n",
       "0      M1L                    S\n",
       "1      M1V                    S\n",
       "2      M1K                    S\n",
       "3      M1T                    S\n",
       "4      M1R                    S"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pandas.DataFrame(mutations, columns=['MUTATION'])\n",
    "\n",
    "# we don't know the phenotypes so arbitratily set them all to susceptible\n",
    "df[['CONSISTENT_PHENOTYPE']] = 'S'\n",
    "\n",
    "# save to disc\n",
    "df.to_csv('data/ds-all-phen.csv', index=False)\n",
    "df[:5]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At this point return to `06-add-features`, set `filestem = 'all'` and run all cells to add all features to this dataset.\n",
    "\n",
    "Now load the best XGBoost model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "best_model={}\n",
    "for model in ['LR', 'NN', 'XB']:\n",
    "    best_model[model] = load('models/'+model.lower()+'.skops', trusted=True)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the dataset with features added"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>segid</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>d_volume</th>\n",
       "      <th>d_hydropathy_KD</th>\n",
       "      <th>d_Pi</th>\n",
       "      <th>d_MW</th>\n",
       "      <th>d_rogov</th>\n",
       "      <th>phi</th>\n",
       "      <th>psi</th>\n",
       "      <th>residue_sasa</th>\n",
       "      <th>n_hbond_acceptors</th>\n",
       "      <th>n_hbond_donors</th>\n",
       "      <th>secondary_structure_codes</th>\n",
       "      <th>temp_factor</th>\n",
       "      <th>depth</th>\n",
       "      <th>deep_ddG</th>\n",
       "      <th>rasp_score_ml</th>\n",
       "      <th>snap2_score</th>\n",
       "      <th>dist_FE2</th>\n",
       "      <th>dist_PZA</th>\n",
       "      <th>mcsm_stability_ddG</th>\n",
       "      <th>mapp_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1I</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "      <td>3.8</td>\n",
       "      <td>2.6</td>\n",
       "      <td>0.28</td>\n",
       "      <td>-18.0</td>\n",
       "      <td>-0.452</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-35.2</td>\n",
       "      <td>86.2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>59.720001</td>\n",
       "      <td>2.103049</td>\n",
       "      <td>-0.407</td>\n",
       "      <td>0.076292</td>\n",
       "      <td>7</td>\n",
       "      <td>29.948318</td>\n",
       "      <td>27.641663</td>\n",
       "      <td>-0.771</td>\n",
       "      <td>21.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1K</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "      <td>5.7</td>\n",
       "      <td>-5.8</td>\n",
       "      <td>4.00</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>-0.712</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-35.2</td>\n",
       "      <td>86.2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>59.720001</td>\n",
       "      <td>2.103049</td>\n",
       "      <td>-1.017</td>\n",
       "      <td>0.621006</td>\n",
       "      <td>70</td>\n",
       "      <td>29.948318</td>\n",
       "      <td>27.641663</td>\n",
       "      <td>-0.214</td>\n",
       "      <td>27.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M1L</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "      <td>3.8</td>\n",
       "      <td>1.9</td>\n",
       "      <td>0.24</td>\n",
       "      <td>-18.0</td>\n",
       "      <td>-0.389</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-35.2</td>\n",
       "      <td>86.2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>59.720001</td>\n",
       "      <td>2.103049</td>\n",
       "      <td>-0.770</td>\n",
       "      <td>-0.107405</td>\n",
       "      <td>20</td>\n",
       "      <td>29.948318</td>\n",
       "      <td>27.641663</td>\n",
       "      <td>-0.771</td>\n",
       "      <td>13.49</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  mutation segid phenotype  d_volume  d_hydropathy_KD  d_Pi  d_MW  d_rogov  \\\n",
       "0      M1I     A         S       3.8              2.6  0.28 -18.0   -0.452   \n",
       "1      M1K     A         S       5.7             -5.8  4.00  -3.0   -0.712   \n",
       "2      M1L     A         S       3.8              1.9  0.24 -18.0   -0.389   \n",
       "\n",
       "   phi   psi  residue_sasa  n_hbond_acceptors  n_hbond_donors  \\\n",
       "0  0.0 -35.2          86.2                2.0             0.0   \n",
       "1  0.0 -35.2          86.2                2.0             0.0   \n",
       "2  0.0 -35.2          86.2                2.0             0.0   \n",
       "\n",
       "   secondary_structure_codes  temp_factor     depth  deep_ddG  rasp_score_ml  \\\n",
       "0                          4    59.720001  2.103049    -0.407       0.076292   \n",
       "1                          4    59.720001  2.103049    -1.017       0.621006   \n",
       "2                          4    59.720001  2.103049    -0.770      -0.107405   \n",
       "\n",
       "   snap2_score   dist_FE2   dist_PZA  mcsm_stability_ddG  mapp_score  \n",
       "0            7  29.948318  27.641663              -0.771       21.99  \n",
       "1           70  29.948318  27.641663              -0.214       27.72  \n",
       "2           20  29.948318  27.641663              -0.771       13.49  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pandas.read_csv('data/ds-all-phen-features.csv')\n",
    "df[:3]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the pipeline since we will need to transform the features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "pipe = joblib.load('data/pipeline.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations ={}\n",
    "df = pandas.read_csv('data/ds-all-phen-features.csv')\n",
    "df.drop(columns=['secondary_structure_codes',\n",
    "    'd_MW',\n",
    "    'phi',\n",
    "    'd_volume',\n",
    "    'd_Pi',\n",
    "    'n_hbond_acceptors'],inplace=True)\n",
    "# make one of the mutations resistant to force it to calculate the statistics\n",
    "df.loc[(df.mutation=='M1L'), 'phenotype']='R'\n",
    "\n",
    "mutations['v'] = df['mutation']\n",
    "features = df.iloc[:,3:]\n",
    "labels = df['phenotype'].map({'S':0, 'R':1}) #, 'U':2})\n",
    "\n",
    "X = pipe.transform(features)\n",
    "Y = labels.squeeze().to_numpy()\n",
    "Z = mutations['v'].to_numpy()\n",
    "\n",
    "with open('data/ds-all.npy', 'wb') as f:\n",
    "    numpy.save(f, Y)\n",
    "    numpy.save(f, X)\n",
    "    numpy.save(f, Z)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "X={}\n",
    "Y={}\n",
    "Z={}\n",
    "\n",
    "for i in ['all']:\n",
    "    X[i]={}\n",
    "    Y[i]={}\n",
    "    Z[i]={}\n",
    "    with open('data/ds-'+i+'.npy', 'rb') as f:\n",
    "        Y[i]['input'] = numpy.load(f)\n",
    "        X[i]['input'] = numpy.load(f)\n",
    "        Z[i]['input'] = numpy.load(f, allow_pickle=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>dataset</th>\n",
       "      <th>sensitivity_mean</th>\n",
       "      <th>sensitivity_std</th>\n",
       "      <th>specificity_mean</th>\n",
       "      <th>specificity_std</th>\n",
       "      <th>roc_auc_mean</th>\n",
       "      <th>roc_auc_std</th>\n",
       "      <th>TN</th>\n",
       "      <th>FP</th>\n",
       "      <th>FN</th>\n",
       "      <th>TP</th>\n",
       "      <th>model_parameters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LR</td>\n",
       "      <td>all</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>52.823315</td>\n",
       "      <td>None</td>\n",
       "      <td>31.785064</td>\n",
       "      <td>None</td>\n",
       "      <td>580</td>\n",
       "      <td>518</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>null</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NN</td>\n",
       "      <td>all</td>\n",
       "      <td>100.0</td>\n",
       "      <td>None</td>\n",
       "      <td>47.905282</td>\n",
       "      <td>None</td>\n",
       "      <td>62.750455</td>\n",
       "      <td>None</td>\n",
       "      <td>526</td>\n",
       "      <td>572</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>null</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XB</td>\n",
       "      <td>all</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>53.278689</td>\n",
       "      <td>None</td>\n",
       "      <td>52.641166</td>\n",
       "      <td>None</td>\n",
       "      <td>585</td>\n",
       "      <td>513</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>null</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  model dataset  sensitivity_mean sensitivity_std  specificity_mean  \\\n",
       "0    LR     all               0.0            None         52.823315   \n",
       "1    NN     all             100.0            None         47.905282   \n",
       "2    XB     all               0.0            None         53.278689   \n",
       "\n",
       "  specificity_std  roc_auc_mean roc_auc_std   TN   FP  FN  TP model_parameters  \n",
       "0            None     31.785064        None  580  518   1   0             null  \n",
       "1            None     62.750455        None  526  572   0   1             null  \n",
       "2            None     52.641166        None  585  513   1   0             null  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def validate_model(line, best_model, model_name, X, Y):\n",
    "\n",
    "    for dataset in ['all']: \n",
    "        \n",
    "        Y[dataset]['predicted'] = best_model[model_name].predict(X[dataset]['input'])\n",
    "        Y[dataset]['scores'] = best_model[model_name].predict_proba(X[dataset]['input'])[:,1]\n",
    "\n",
    "        row = construct_line(model_name , dataset, None, Y[dataset], None)\n",
    "        line.append(row)\n",
    "\n",
    "    return(line)\n",
    "\n",
    "line = []\n",
    "for model in ['LR', 'NN', 'XB']:\n",
    "    line = validate_model(line, best_model, model, X, Y)\n",
    "\n",
    "results = pandas.DataFrame(line, columns=['model', 'dataset', 'sensitivity_mean', 'sensitivity_std', 'specificity_mean', 'specificity_std' ,'roc_auc_mean', 'roc_auc_std','TN','FP','FN','TP', 'model_parameters'])\n",
    "results  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
